// react-virtuoso/dist/index.mjs
import { jsx as G, jsxs as Dt } from "react/jsx-runtime";
import E, { createElement as Nt } from "react";
import fo from "react-dom";
const Ce = 0, Pt = 1, Zt = 2, kn = 4;
function an(t) {
  return () => t;
}
function mo(t) {
  t();
}
function oe(t, e) {
  return (n) => t(e(n));
}
function dn(t, e) {
  return () => t(e);
}
function po(t, e) {
  return (n) => t(e, n);
}
function Ae(t) {
  return t !== void 0;
}
function ho(...t) {
  return () => {
    t.map(mo);
  };
}
function Xt() {
}
function we(t, e) {
  return e(t), t;
}
function go(t, e) {
  return e(t);
}
function rt(...t) {
  return t;
}
function Y(t, e) {
  return t(Pt, e);
}
function _(t, e) {
  t(Ce, e);
}
function Me(t) {
  t(Zt);
}
function it(t) {
  return t(kn);
}
function k(t, e) {
  return Y(t, po(e, Ce));
}
function yt(t, e) {
  const n = t(Pt, (o) => {
    n(), e(o);
  });
  return n;
}
function fn(t) {
  let e, n;
  return (o) => (r) => {
    e = r, n && clearTimeout(n), n = setTimeout(() => {
      o(e);
    }, t);
  };
}
function zn(t, e) {
  return t === e;
}
function nt(t = zn) {
  let e;
  return (n) => (o) => {
    t(e, o) || (e = o, n(o));
  };
}
function P(t) {
  return (e) => (n) => {
    t(n) && e(n);
  };
}
function B(t) {
  return (e) => oe(e, t);
}
function Bt(t) {
  return (e) => () => {
    e(t);
  };
}
function I(t, ...e) {
  const n = Io(...e);
  return ((o, r) => {
    switch (o) {
      case Zt:
        Me(t);
        return;
      case Pt:
        return Y(t, n(r));
    }
  });
}
function Ot(t, e) {
  return (n) => (o) => {
    n(e = t(e, o));
  };
}
function $t(t) {
  return (e) => (n) => {
    t > 0 ? t-- : e(n);
  };
}
function zt(t) {
  let e = null, n;
  return (o) => (r) => {
    e = r, !n && (n = setTimeout(() => {
      n = void 0, o(e);
    }, t));
  };
}
function $(...t) {
  const e = new Array(t.length);
  let n = 0, o = null;
  const r = 2 ** t.length - 1;
  return t.forEach((s, i) => {
    const l = 2 ** i;
    Y(s, (c) => {
      const d = n;
      n |= l, e[i] = c, d !== r && n === r && o && (o(), o = null);
    });
  }), (s) => (i) => {
    const l = () => {
      s([i].concat(e));
    };
    n === r ? l() : o = l;
  };
}
function Io(...t) {
  return (e) => t.reduceRight(go, e);
}
function So(t) {
  let e, n;
  const o = () => e?.();
  return function(r, s) {
    switch (r) {
      case Pt:
        return s ? n === s ? void 0 : (o(), n = s, e = Y(t, s), e) : (o(), Xt);
      case Zt:
        o(), n = null;
        return;
    }
  };
}
function v(t) {
  let e = t;
  const n = U();
  return ((o, r) => {
    switch (o) {
      case Ce:
        e = r;
        break;
      case Pt: {
        r(e);
        break;
      }
      case kn:
        return e;
    }
    return n(o, r);
  });
}
function ht(t, e) {
  return we(v(e), (n) => k(t, n));
}
function U() {
  const t = [];
  return ((e, n) => {
    switch (e) {
      case Ce:
        t.slice().forEach((o) => {
          o(n);
        });
        return;
      case Zt:
        t.splice(0, t.length);
        return;
      case Pt:
        return t.push(n), () => {
          const o = t.indexOf(n);
          o > -1 && t.splice(o, 1);
        };
    }
  });
}
function vt(t) {
  return we(U(), (e) => k(t, e));
}
function j(t, e = [], { singleton: n } = { singleton: !0 }) {
  return {
    constructor: t,
    dependencies: e,
    id: xo(),
    singleton: n
  };
}
const xo = () => Symbol();
function To(t) {
  const e = /* @__PURE__ */ new Map(), n = ({ constructor: o, dependencies: r, id: s, singleton: i }) => {
    if (i && e.has(s))
      return e.get(s);
    const l = o(r.map((c) => n(c)));
    return i && e.set(s, l), l;
  };
  return n(t);
}
function at(...t) {
  const e = U(), n = new Array(t.length);
  let o = 0;
  const r = 2 ** t.length - 1;
  return t.forEach((s, i) => {
    const l = 2 ** i;
    Y(s, (c) => {
      n[i] = c, o |= l, o === r && _(e, n);
    });
  }), function(s, i) {
    switch (s) {
      case Zt: {
        Me(e);
        return;
      }
      case Pt:
        return o === r && i(n), Y(e, i);
    }
  };
}
function W(t, e = zn) {
  return I(t, nt(e));
}
function ze(...t) {
  return function(e, n) {
    switch (e) {
      case Zt:
        return;
      case Pt:
        return ho(...t.map((o) => Y(o, n)));
    }
  };
}
const ft = {
  /** Detailed debugging information including item measurements */
  DEBUG: 0,
  /** General informational messages */
  INFO: 1,
  /** Warning messages for potential issues */
  WARN: 2,
  /** Error messages for failures (default level) */
  ERROR: 3
}, vo = {
  [ft.DEBUG]: "debug",
  [ft.ERROR]: "error",
  [ft.INFO]: "log",
  [ft.WARN]: "warn"
}, Co = () => typeof globalThis > "u" ? window : globalThis, Gt = j(
  () => {
    const t = v(ft.ERROR);
    return {
      log: v((n, o, r = ft.INFO) => {
        const s = Co().VIRTUOSO_LOG_LEVEL ?? it(t);
        r >= s && console[vo[r]](
          "%creact-virtuoso: %c%s %o",
          "color: #0253b3; font-weight: bold",
          "color: initial",
          n,
          o
        );
      }),
      logLevel: t
    };
  },
  [],
  { singleton: !0 }
);
function Lt(t, e, n) {
  return _e(t, e, n).callbackRef;
}
function _e(t, e, n) {
  const o = E.useRef(null);
  let r = (i) => {
  };
  const s = E.useMemo(() => typeof ResizeObserver < "u" ? new ResizeObserver((i) => {
    const l = () => {
      const c = i[0].target;
      c.offsetParent !== null && t(c);
    };
    n ? l() : requestAnimationFrame(l);
  }) : null, [t, n]);
  return r = (i) => {
    i && e ? (s?.observe(i), o.current = i) : (o.current && s?.unobserve(o.current), o.current = null);
  }, { callbackRef: r, ref: o };
}
function Fn(t, e, n, o, r, s, i, l, c) {
  const d = E.useCallback(
    (m) => {
      const S = wo(m.children, e, l ? "offsetWidth" : "offsetHeight", r);
      let p = m.parentElement;
      for (; p.dataset.virtuosoScroller === void 0; )
        p = p.parentElement;
      const T = p.lastElementChild.dataset.viewportType === "window";
      let w;
      T && (w = p.ownerDocument.defaultView);
      const R = i ? l ? i.scrollLeft : i.scrollTop : T ? l ? w.scrollX || w.document.documentElement.scrollLeft : w.scrollY || w.document.documentElement.scrollTop : l ? p.scrollLeft : p.scrollTop, h = i ? l ? i.scrollWidth : i.scrollHeight : T ? l ? w.document.documentElement.scrollWidth : w.document.documentElement.scrollHeight : l ? p.scrollWidth : p.scrollHeight, f = i ? l ? i.offsetWidth : i.offsetHeight : T ? l ? w.innerWidth : w.innerHeight : l ? p.offsetWidth : p.offsetHeight;
      o({
        scrollHeight: h,
        scrollTop: Math.max(R, 0),
        viewportHeight: f
      }), s?.(
        l ? mn("column-gap", getComputedStyle(m).columnGap, r) : mn("row-gap", getComputedStyle(m).rowGap, r)
      ), S !== null && t(S);
    },
    [t, e, r, s, i, o, l]
  );
  return _e(d, n, c);
}
function wo(t, e, n, o) {
  const r = t.length;
  if (r === 0)
    return null;
  const s = [];
  for (let i = 0; i < r; i++) {
    const l = t.item(i);
    if (l.dataset.index === void 0)
      continue;
    const c = parseInt(l.dataset.index), d = parseFloat(l.dataset.knownSize), m = e(l, n);
    if (m === 0 && o("Zero-sized element, this should not happen", { child: l }, ft.ERROR), m === d)
      continue;
    const S = s[s.length - 1];
    s.length === 0 || S.size !== m || S.endIndex !== c - 1 ? s.push({ endIndex: c, size: m, startIndex: c }) : s[s.length - 1].endIndex++;
  }
  return s;
}
function mn(t, e, n) {
  return e !== "normal" && e?.endsWith("px") !== !0 && n(`${t} was not resolved to pixel value correctly`, e, ft.WARN), e === "normal" ? 0 : parseInt(e ?? "0", 10);
}
function Ne(t, e, n) {
  const o = E.useRef(null), r = E.useCallback(
    (c) => {
      if (!c?.offsetParent)
        return;
      const d = c.getBoundingClientRect(), m = d.width;
      let S, p;
      if (e) {
        const T = e.getBoundingClientRect(), w = d.top - T.top;
        p = T.height - Math.max(0, w), S = w + e.scrollTop;
      } else {
        const T = i.current.ownerDocument.defaultView;
        p = T.innerHeight - Math.max(0, d.top), S = d.top + T.scrollY;
      }
      o.current = {
        listHeight: d.height,
        offsetTop: S,
        visibleHeight: p,
        visibleWidth: m
      }, t(o.current);
    },
    // oxlint-disable-next-line exhaustive-deps
    [t, e]
  ), { callbackRef: s, ref: i } = _e(r, !0, n), l = E.useCallback(() => {
    r(i.current);
  }, [r, i]);
  return E.useEffect(() => {
    if (e) {
      e.addEventListener("scroll", l);
      const d = new ResizeObserver(() => {
        requestAnimationFrame(l);
      });
      return d.observe(e), () => {
        e.removeEventListener("scroll", l), d.unobserve(e);
      };
    }
    const c = i.current?.ownerDocument.defaultView;
    return c?.addEventListener("scroll", l), c?.addEventListener("resize", l), () => {
      c?.removeEventListener("scroll", l), c?.removeEventListener("resize", l);
    };
  }, [l, e, i]), s;
}
const It = j(
  () => {
    const t = U(), e = U(), n = v(0), o = U(), r = v(0), s = U(), i = U(), l = v(0), c = v(0), d = v(0), m = v(0), S = U(), p = U(), T = v(!1), w = v(!1), R = v(!1);
    return k(
      I(
        t,
        B(({ scrollTop: h }) => h)
      ),
      e
    ), k(
      I(
        t,
        B(({ scrollHeight: h }) => h)
      ),
      i
    ), k(e, r), {
      deviation: n,
      fixedFooterHeight: d,
      fixedHeaderHeight: c,
      footerHeight: m,
      headerHeight: l,
      horizontalDirection: w,
      scrollBy: p,
      // input
      scrollContainerState: t,
      scrollHeight: i,
      scrollingInProgress: T,
      // signals
      scrollTo: S,
      scrollTop: e,
      skipAnimationFrameInResizeObserver: R,
      smoothScrollTargetReached: o,
      // state
      statefulScrollTop: r,
      viewportHeight: s
    };
  },
  [],
  { singleton: !0 }
), re = { lvl: 0 };
function Vn(t, e) {
  const n = t.length;
  if (n === 0)
    return [];
  let { index: o, value: r } = e(t[0]);
  const s = [];
  for (let i = 1; i < n; i++) {
    const { index: l, value: c } = e(t[i]);
    s.push({ end: l - 1, start: o, value: r }), o = l, r = c;
  }
  return s.push({ end: 1 / 0, start: o, value: r }), s;
}
function J(t) {
  return t === re;
}
function se(t, e) {
  if (!J(t))
    return e === t.k ? t.v : e < t.k ? se(t.l, e) : se(t.r, e);
}
function Rt(t, e, n = "k") {
  if (J(t))
    return [-1 / 0, void 0];
  if (Number(t[n]) === e)
    return [t.k, t.v];
  if (Number(t[n]) < e) {
    const o = Rt(t.r, e, n);
    return o[0] === -1 / 0 ? [t.k, t.v] : o;
  }
  return Rt(t.l, e, n);
}
function Tt(t, e, n) {
  return J(t) ? Gn(e, n, 1) : e === t.k ? dt(t, { k: e, v: n }) : e < t.k ? pn(dt(t, { l: Tt(t.l, e, n) })) : pn(dt(t, { r: Tt(t.r, e, n) }));
}
function qt() {
  return re;
}
function Yt(t, e, n) {
  if (J(t))
    return [];
  const o = Rt(t, e)[0];
  return yo(Ve(t, o, n));
}
function Fe(t, e) {
  if (J(t))
    return re;
  const { k: n, l: o, r } = t;
  if (e === n) {
    if (J(o))
      return r;
    if (J(r))
      return o;
    const [s, i] = Pn(o);
    return Se(dt(t, { k: s, l: Wn(o), v: i }));
  }
  return e < n ? Se(dt(t, { l: Fe(o, e) })) : Se(dt(t, { r: Fe(r, e) }));
}
function _t(t) {
  return J(t) ? [] : [..._t(t.l), { k: t.k, v: t.v }, ..._t(t.r)];
}
function Ve(t, e, n) {
  if (J(t))
    return [];
  const { k: o, l: r, r: s, v: i } = t;
  let l = [];
  return o > e && (l = l.concat(Ve(r, e, n))), o >= e && o <= n && l.push({ k: o, v: i }), o <= n && (l = l.concat(Ve(s, e, n))), l;
}
function Se(t) {
  const { l: e, lvl: n, r: o } = t;
  if (o.lvl >= n - 1 && e.lvl >= n - 1)
    return t;
  if (n > o.lvl + 1) {
    if (Ee(e))
      return An(dt(t, { lvl: n - 1 }));
    if (!J(e) && !J(e.r))
      return dt(e.r, {
        l: dt(e, { r: e.r.l }),
        lvl: n,
        r: dt(t, {
          l: e.r.r,
          lvl: n - 1
        })
      });
    throw new Error("Unexpected empty nodes");
  }
  if (Ee(t))
    return We(dt(t, { lvl: n - 1 }));
  if (!J(o) && !J(o.l)) {
    const r = o.l, s = Ee(r) ? o.lvl - 1 : o.lvl;
    return dt(r, {
      l: dt(t, {
        lvl: n - 1,
        r: r.l
      }),
      lvl: r.lvl + 1,
      r: We(dt(o, { l: r.r, lvl: s }))
    });
  }
  throw new Error("Unexpected empty nodes");
}
function dt(t, e) {
  return Gn(
    e.k !== void 0 ? e.k : t.k,
    e.v !== void 0 ? e.v : t.v,
    e.lvl !== void 0 ? e.lvl : t.lvl,
    e.l !== void 0 ? e.l : t.l,
    e.r !== void 0 ? e.r : t.r
  );
}
function Wn(t) {
  return J(t.r) ? t.l : Se(dt(t, { r: Wn(t.r) }));
}
function Ee(t) {
  return J(t) || t.lvl > t.r.lvl;
}
function Pn(t) {
  return J(t.r) ? [t.k, t.v] : Pn(t.r);
}
function Gn(t, e, n, o = re, r = re) {
  return { k: t, l: o, lvl: n, r, v: e };
}
function pn(t) {
  return We(An(t));
}
function An(t) {
  const { l: e } = t;
  return !J(e) && e.lvl === t.lvl ? dt(e, { r: dt(t, { l: e.r }) }) : t;
}
function We(t) {
  const { lvl: e, r: n } = t;
  return !J(n) && !J(n.r) && n.lvl === e && n.r.lvl === e ? dt(n, { l: dt(t, { r: n.l }), lvl: e + 1 }) : t;
}
function yo(t) {
  return Vn(t, ({ k: e, v: n }) => ({ index: e, value: n }));
}
function Mn(t, e) {
  return !!(t && t.startIndex === e.startIndex && t.endIndex === e.endIndex);
}
function ie(t, e) {
  return !!(t && t[0] === e[0] && t[1] === e[1]);
}
const De = j(
  () => ({ recalcInProgress: v(!1) }),
  [],
  { singleton: !0 }
);
function _n(t, e, n) {
  return t[Te(t, e, n)];
}
function Te(t, e, n, o = 0) {
  let r = t.length - 1;
  for (; o <= r; ) {
    const s = Math.floor((o + r) / 2), i = t[s], l = n(i, e);
    if (l === 0)
      return s;
    if (l === -1) {
      if (r - o < 2)
        return s - 1;
      r = s - 1;
    } else {
      if (r === o)
        return s;
      o = s + 1;
    }
  }
  throw new Error(`Failed binary finding record in array - ${t.join(",")}, searched for ${e}`);
}
function bo(t, e, n, o) {
  const r = Te(t, e, o), s = Te(t, n, o, r);
  return t.slice(r, s + 1);
}
function Ht(t, e) {
  return Math.round(t.getBoundingClientRect()[e]);
}
function ye(t) {
  return !J(t.groupOffsetTree);
}
function $e({ index: t }, e) {
  return e === t ? 0 : e < t ? -1 : 1;
}
function Ro() {
  return {
    groupIndices: [],
    groupOffsetTree: qt(),
    lastIndex: 0,
    lastOffset: 0,
    lastSize: 0,
    offsetTree: [],
    sizeTree: qt()
  };
}
function Ho(t, e) {
  let n = J(t) ? 0 : 1 / 0;
  for (const o of e) {
    const { endIndex: r, size: s, startIndex: i } = o;
    if (n = Math.min(n, i), J(t)) {
      t = Tt(t, 0, s);
      continue;
    }
    const l = Yt(t, i - 1, r + 1);
    if (l.some(Fo(o)))
      continue;
    let c = !1, d = !1;
    for (const { end: m, start: S, value: p } of l)
      c ? (r >= S || s === p) && (t = Fe(t, S)) : (d = p !== s, c = !0), m > r && r >= S && p !== s && (t = Tt(t, r + 1, p));
    d && (t = Tt(t, i, s));
  }
  return [t, n];
}
function Eo(t) {
  return typeof t.groupIndex < "u";
}
function Bo({ offset: t }, e) {
  return e === t ? 0 : e < t ? -1 : 1;
}
function le(t, e, n) {
  if (e.length === 0)
    return 0;
  const { index: o, offset: r, size: s } = _n(e, t, $e), i = t - o, l = s * i + (i - 1) * n + r;
  return l > 0 ? l + n : l;
}
function Nn(t, e) {
  if (!ye(e))
    return t;
  let n = 0;
  for (; e.groupIndices[n] <= t + n; )
    n++;
  return t + n;
}
function Dn(t, e, n) {
  if (Eo(t))
    return e.groupIndices[t.groupIndex] + 1;
  const o = t.index === "LAST" ? n : t.index;
  let r = Nn(o, e);
  return r = Math.max(0, r, Math.min(n, r)), r;
}
function Oo(t, e, n, o = 0) {
  return o > 0 && (e = Math.max(e, _n(t, o, $e).offset)), Vn(bo(t, e, n, Bo), zo);
}
function Lo(t, [e, n, o, r]) {
  e.length > 0 && o("received item sizes", e, ft.DEBUG);
  const s = t.sizeTree;
  let i = s, l = 0;
  if (n.length > 0 && J(s) && e.length === 2) {
    const p = e[0].size, T = e[1].size;
    i = n.reduce((w, R) => Tt(Tt(w, R, p), R + 1, T), i);
  } else
    [i, l] = Ho(i, e);
  if (i === s)
    return t;
  const { lastIndex: c, lastOffset: d, lastSize: m, offsetTree: S } = Pe(t.offsetTree, l, i, r);
  return {
    groupIndices: n,
    groupOffsetTree: n.reduce((p, T) => Tt(p, T, le(T, S, r)), qt()),
    lastIndex: c,
    lastOffset: d,
    lastSize: m,
    offsetTree: S,
    sizeTree: i
  };
}
function ko(t) {
  return _t(t).map(({ k: e, v: n }, o, r) => {
    const s = r[o + 1];
    return { endIndex: s !== void 0 ? s.k - 1 : 1 / 0, size: n, startIndex: e };
  });
}
function hn(t, e) {
  let n = 0, o = 0;
  for (; n < t; )
    n += e[o + 1] - e[o] - 1, o++;
  return o - (n === t ? 0 : 1);
}
function Pe(t, e, n, o) {
  let r = t, s = 0, i = 0, l = 0, c = 0;
  if (e !== 0) {
    c = Te(r, e - 1, $e), l = r[c].offset;
    const m = Rt(n, e - 1);
    s = m[0], i = m[1], r.length && r[c].size === Rt(n, e)[1] && (c -= 1), r = r.slice(0, c + 1);
  } else
    r = [];
  for (const { start: d, value: m } of Yt(n, e, 1 / 0)) {
    const S = d - s, p = S * i + l + S * o;
    r.push({
      index: d,
      offset: p,
      size: m
    }), s = d, l = p, i = m;
  }
  return {
    lastIndex: s,
    lastOffset: l,
    lastSize: i,
    offsetTree: r
  };
}
function zo(t) {
  return { index: t.index, value: t };
}
function Fo(t) {
  const { endIndex: e, size: n, startIndex: o } = t;
  return (r) => r.start === o && (r.end === e || r.end === 1 / 0) && r.value === n;
}
const Vo = {
  offsetHeight: "height",
  offsetWidth: "width"
}, kt = j(
  ([{ log: t }, { recalcInProgress: e }]) => {
    const n = U(), o = U(), r = ht(o, 0), s = U(), i = U(), l = v(0), c = v([]), d = v(void 0), m = v(void 0), S = v(void 0), p = v(void 0), T = v((u, g) => Ht(u, Vo[g])), w = v(void 0), R = v(0), h = Ro(), f = ht(
      I(n, $(c, t, R), Ot(Lo, h), nt()),
      h
    ), a = ht(
      I(
        c,
        nt(),
        Ot((u, g) => ({ current: g, prev: u.current }), {
          current: [],
          prev: []
        }),
        B(({ prev: u }) => u)
      ),
      []
    );
    k(
      I(
        c,
        P((u) => u.length > 0),
        $(f, R),
        B(([u, g, C]) => {
          const z = u.reduce((O, V, N) => Tt(O, V, le(V, g.offsetTree, C) || N), qt());
          return {
            ...g,
            groupIndices: u,
            groupOffsetTree: z
          };
        })
      ),
      f
    ), k(
      I(
        o,
        $(f),
        P(([u, { lastIndex: g }]) => u < g),
        B(([u, { lastIndex: g, lastSize: C }]) => [
          {
            endIndex: g,
            size: C,
            startIndex: u
          }
        ])
      ),
      n
    ), k(d, m);
    const x = ht(
      I(
        d,
        B((u) => u === void 0)
      ),
      !0
    );
    k(
      I(
        m,
        P((u) => u !== void 0 && J(it(f).sizeTree)),
        B((u) => {
          const g = it(S), C = it(c).length > 0;
          return g !== void 0 && g !== 0 ? C ? [
            { endIndex: 0, size: g, startIndex: 0 },
            { endIndex: 1, size: u, startIndex: 1 }
          ] : [] : [{ endIndex: 0, size: u, startIndex: 0 }];
        })
      ),
      n
    ), k(
      I(
        p,
        P((u) => u !== void 0 && u.length > 0 && J(it(f).sizeTree)),
        B((u) => {
          const g = [];
          let C = u[0], z = 0;
          for (let O = 1; O < u.length; O++) {
            const V = u[O];
            V !== C && (g.push({
              endIndex: O - 1,
              size: C,
              startIndex: z
            }), C = V, z = O);
          }
          return g.push({
            endIndex: u.length - 1,
            size: C,
            startIndex: z
          }), g;
        })
      ),
      n
    ), k(
      I(
        c,
        $(S, m),
        P(([, u, g]) => u !== void 0 && g !== void 0),
        B(([u, g, C]) => {
          const z = [];
          for (let O = 0; O < u.length; O++) {
            const V = u[O], N = u[O + 1];
            z.push({
              startIndex: V,
              endIndex: V,
              size: g
            }), N !== void 0 && z.push({
              startIndex: V + 1,
              endIndex: N - 1,
              size: C
            });
          }
          return z;
        })
      ),
      n
    );
    const H = vt(
      I(
        n,
        $(f),
        Ot(
          ({ sizes: u }, [g, C]) => ({
            changed: C !== u,
            sizes: C
          }),
          { changed: !1, sizes: h }
        ),
        B((u) => u.changed)
      )
    );
    Y(
      I(
        l,
        Ot(
          (u, g) => ({ diff: u.prev - g, prev: g }),
          { diff: 0, prev: 0 }
        ),
        B((u) => u.diff)
      ),
      (u) => {
        const { groupIndices: g } = it(f);
        if (u > 0)
          _(e, !0), _(s, u + hn(u, g));
        else if (u < 0) {
          const C = it(a);
          C.length > 0 && (u -= hn(-u, C)), _(i, u);
        }
      }
    ), Y(I(l, $(t)), ([u, g]) => {
      u < 0 && g(
        "`firstItemIndex` prop should not be set to less than zero. If you don't know the total count, just use a very high value",
        { firstItemIndex: l },
        ft.ERROR
      );
    });
    const y = vt(s);
    k(
      I(
        s,
        $(f),
        B(([u, g]) => {
          const C = g.groupIndices.length > 0, z = [], O = g.lastSize;
          if (C) {
            const V = se(g.sizeTree, 0);
            let N = 0, Z = 0;
            for (; N < u; ) {
              const q = g.groupIndices[Z], Q = g.groupIndices.length === Z + 1 ? 1 / 0 : g.groupIndices[Z + 1] - q - 1;
              z.push({
                endIndex: q,
                size: V,
                startIndex: q
              }), z.push({
                endIndex: q + 1 + Q - 1,
                size: O,
                startIndex: q + 1
              }), Z++, N += Q + 1;
            }
            const F = _t(g.sizeTree);
            return N !== u && F.shift(), F.reduce(
              (q, { k: Q, v: gt }) => {
                let ut = q.ranges;
                return q.prevSize !== 0 && (ut = [
                  ...q.ranges,
                  {
                    endIndex: Q + u - 1,
                    size: q.prevSize,
                    startIndex: q.prevIndex
                  }
                ]), {
                  prevIndex: Q + u,
                  prevSize: gt,
                  ranges: ut
                };
              },
              {
                prevIndex: u,
                prevSize: 0,
                ranges: z
              }
            ).ranges;
          }
          return _t(g.sizeTree).reduce(
            (V, { k: N, v: Z }) => ({
              prevIndex: N + u,
              prevSize: Z,
              ranges: [...V.ranges, { endIndex: N + u - 1, size: V.prevSize, startIndex: V.prevIndex }]
            }),
            {
              prevIndex: 0,
              prevSize: O,
              ranges: []
            }
          ).ranges;
        })
      ),
      n
    );
    const L = vt(
      I(
        i,
        $(f, R),
        B(([u, { offsetTree: g }, C]) => {
          const z = -u;
          return le(z, g, C);
        })
      )
    );
    return k(
      I(
        i,
        $(f, R),
        B(([u, g, C]) => {
          if (g.groupIndices.length > 0) {
            if (J(g.sizeTree))
              return g;
            let V = qt();
            const N = it(a);
            let Z = 0, F = 0, mt = 0;
            for (; Z < -u; ) {
              mt = N[F];
              const Q = N[F + 1] - mt - 1;
              F++, Z += Q + 1;
            }
            if (V = _t(g.sizeTree).reduce((Q, { k: gt, v: ut }) => Tt(Q, Math.max(0, gt + u), ut), V), Z !== -u) {
              const Q = se(g.sizeTree, mt);
              V = Tt(V, 0, Q);
              const gt = Rt(g.sizeTree, -u + 1)[1];
              V = Tt(V, 1, gt);
            }
            return {
              ...g,
              sizeTree: V,
              ...Pe(g.offsetTree, 0, V, C)
            };
          }
          const O = _t(g.sizeTree).reduce((V, { k: N, v: Z }) => Tt(V, Math.max(0, N + u), Z), qt());
          return {
            ...g,
            sizeTree: O,
            ...Pe(g.offsetTree, 0, O, C)
          };
        })
      ),
      f
    ), {
      beforeUnshiftWith: y,
      // input
      data: w,
      defaultItemSize: m,
      firstItemIndex: l,
      fixedItemSize: d,
      fixedGroupSize: S,
      gap: R,
      groupIndices: c,
      heightEstimates: p,
      itemSize: T,
      listRefresh: H,
      shiftWith: i,
      shiftWithOffset: L,
      sizeRanges: n,
      // output
      sizes: f,
      statefulTotalCount: r,
      totalCount: o,
      trackItemSizes: x,
      unshiftWith: s
    };
  },
  rt(Gt, De),
  { singleton: !0 }
);
function Wo(t) {
  return t.reduce(
    (e, n) => (e.groupIndices.push(e.totalCount), e.totalCount += n + 1, e),
    {
      groupIndices: [],
      totalCount: 0
    }
  );
}
const $n = j(
  ([{ groupIndices: t, sizes: e, totalCount: n }, { headerHeight: o, scrollTop: r }]) => {
    const s = U(), i = U(), l = vt(I(s, B(Wo)));
    return k(
      I(
        l,
        B((c) => c.totalCount)
      ),
      n
    ), k(
      I(
        l,
        B((c) => c.groupIndices)
      ),
      t
    ), k(
      I(
        at(r, e, o),
        P(([c, d]) => ye(d)),
        B(([c, d, m]) => Rt(d.groupOffsetTree, Math.max(c - m, 0), "v")[0]),
        nt(),
        B((c) => [c])
      ),
      i
    ), { groupCounts: s, topItemsIndexes: i };
  },
  rt(kt, It)
), At = j(
  ([{ log: t }]) => {
    const e = v(!1), n = vt(
      I(
        e,
        P((o) => o),
        nt()
      )
    );
    return Y(e, (o) => {
      o && it(t)("props updated", {}, ft.DEBUG);
    }), { didMount: n, propsReady: e };
  },
  rt(Gt),
  { singleton: !0 }
), Po = typeof document < "u" && "scrollBehavior" in document.documentElement.style;
function Un(t) {
  const e = typeof t == "number" ? { index: t } : t;
  return e.align || (e.align = "start"), (!e.behavior || !Po) && (e.behavior = "auto"), e.offset === void 0 && (e.offset = 0), e;
}
const de = j(
  ([
    { gap: t, listRefresh: e, sizes: n, totalCount: o },
    {
      fixedFooterHeight: r,
      fixedHeaderHeight: s,
      footerHeight: i,
      headerHeight: l,
      scrollingInProgress: c,
      scrollTo: d,
      smoothScrollTargetReached: m,
      viewportHeight: S
    },
    { log: p }
  ]) => {
    const T = U(), w = U(), R = v(0);
    let h = null, f = null, a = null;
    function x() {
      h !== null && (h(), h = null), a !== null && (a(), a = null), f && (clearTimeout(f), f = null), _(c, !1);
    }
    return k(
      I(
        T,
        $(n, S, o, R, l, i, p),
        $(t, s, r),
        B(
          ([
            [H, y, L, u, g, C, z, O],
            V,
            N,
            Z
          ]) => {
            const F = Un(H), { align: mt, behavior: q, offset: Q } = F, gt = u - 1, ut = Dn(F, y, gt);
            let St = le(ut, y.offsetTree, V) + C;
            mt === "end" ? (St += N + Rt(y.sizeTree, ut)[1] - L + Z, ut === gt && (St += z)) : mt === "center" ? St += (N + Rt(y.sizeTree, ut)[1] - L + Z) / 2 : St -= g, Q !== void 0 && Q !== 0 && (St += Q);
            const Ft = (pt) => {
              x(), pt ? (O("retrying to scroll to", { location: H }, ft.DEBUG), _(T, H)) : (_(w, !0), O("list did not change, scroll successful", {}, ft.DEBUG));
            };
            if (x(), q === "smooth") {
              let pt = !1;
              a = Y(e, (Kt) => {
                pt = pt || Kt;
              }), h = yt(m, () => {
                Ft(pt);
              });
            } else
              h = yt(I(e, Go(150)), Ft);
            return f = setTimeout(() => {
              x();
            }, 1200), _(c, !0), O("scrolling from index to", { behavior: q, index: ut, top: St }, ft.DEBUG), { behavior: q, top: St };
          }
        )
      ),
      d
    ), {
      scrollTargetReached: w,
      scrollToIndex: T,
      topListHeight: R
    };
  },
  rt(kt, It, Gt),
  { singleton: !0 }
);
function Go(t) {
  return (e) => {
    const n = setTimeout(() => {
      e(!1);
    }, t);
    return (o) => {
      o && (e(!0), clearTimeout(n));
    };
  };
}
function Ue(t, e) {
  t === 0 ? e() : requestAnimationFrame(() => {
    Ue(t - 1, e);
  });
}
function Ke(t, e) {
  const n = e - 1;
  return typeof t == "number" ? t : t.index === "LAST" ? n : t.index;
}
const fe = j(
  ([{ defaultItemSize: t, listRefresh: e, sizes: n }, { scrollTop: o }, { scrollTargetReached: r, scrollToIndex: s }, { didMount: i }]) => {
    const l = v(!0), c = v(0), d = v(!0);
    return k(
      I(
        i,
        $(c),
        P(([m, S]) => S !== 0),
        Bt(!1)
      ),
      l
    ), k(
      I(
        i,
        $(c),
        P(([m, S]) => S !== 0),
        Bt(!1)
      ),
      d
    ), Y(
      I(
        at(e, i),
        $(l, n, t, d),
        P(([[, m], S, { sizeTree: p }, T, w]) => m && (!J(p) || Ae(T)) && !S && !w),
        $(c)
      ),
      ([, m]) => {
        yt(r, () => {
          _(d, !0);
        }), Ue(4, () => {
          yt(o, () => {
            _(l, !0);
          }), _(s, m);
        });
      }
    ), {
      initialItemFinalLocationReached: d,
      initialTopMostItemIndex: c,
      scrolledToInitialItem: l
    };
  },
  rt(kt, It, de, At),
  { singleton: !0 }
);
function Kn(t, e) {
  return Math.abs(t - e) < 1.01;
}
const ce = "up", ee = "down", Ao = "none", Mo = {
  atBottom: !1,
  notAtBottomBecause: "NOT_SHOWING_LAST_ITEM",
  state: {
    offsetBottom: 0,
    scrollHeight: 0,
    scrollTop: 0,
    viewportHeight: 0
  }
}, _o = 0, me = j(([{ footerHeight: t, headerHeight: e, scrollBy: n, scrollContainerState: o, scrollTop: r, viewportHeight: s }]) => {
  const i = v(!1), l = v(!0), c = U(), d = U(), m = v(4), S = v(_o), p = ht(
    I(
      ze(I(W(r), $t(1), Bt(!0)), I(W(r), $t(1), Bt(!1), fn(100))),
      nt()
    ),
    !1
  ), T = ht(
    I(ze(I(n, Bt(!0)), I(n, Bt(!1), fn(200))), nt()),
    !1
  );
  k(
    I(
      at(W(r), W(S)),
      B(([a, x]) => a <= x),
      nt()
    ),
    l
  ), k(I(l, zt(50)), d);
  const w = vt(
    I(
      at(o, W(s), W(e), W(t), W(m)),
      Ot((a, [{ scrollHeight: x, scrollTop: H }, y, L, u, g]) => {
        const C = H + y - x > -g, z = {
          scrollHeight: x,
          scrollTop: H,
          viewportHeight: y
        };
        if (C) {
          let V, N;
          return H > a.state.scrollTop ? (V = "SCROLLED_DOWN", N = a.state.scrollTop - H) : (V = "SIZE_DECREASED", N = a.state.scrollTop - H || a.scrollTopDelta), {
            atBottom: !0,
            atBottomBecause: V,
            scrollTopDelta: N,
            state: z
          };
        }
        let O;
        return z.scrollHeight > a.state.scrollHeight ? O = "SIZE_INCREASED" : y < a.state.viewportHeight ? O = "VIEWPORT_HEIGHT_DECREASING" : H < a.state.scrollTop ? O = "SCROLLING_UPWARDS" : O = "NOT_FULLY_SCROLLED_TO_LAST_ITEM_BOTTOM", {
          atBottom: !1,
          notAtBottomBecause: O,
          state: z
        };
      }, Mo),
      nt((a, x) => a !== void 0 && a.atBottom === x.atBottom)
    )
  ), R = ht(
    I(
      o,
      Ot(
        (a, { scrollHeight: x, scrollTop: H, viewportHeight: y }) => {
          if (!Kn(a.scrollHeight, x)) {
            const L = x - (H + y) < 1;
            return a.scrollTop !== H && L ? {
              changed: !0,
              jump: a.scrollTop - H,
              scrollHeight: x,
              scrollTop: H
            } : {
              changed: !0,
              jump: 0,
              scrollHeight: x,
              scrollTop: H
            };
          }
          return {
            changed: !1,
            jump: 0,
            scrollHeight: x,
            scrollTop: H
          };
        },
        { changed: !1, jump: 0, scrollHeight: 0, scrollTop: 0 }
      ),
      P((a) => a.changed),
      B((a) => a.jump)
    ),
    0
  );
  k(
    I(
      w,
      B((a) => a.atBottom)
    ),
    i
  ), k(I(i, zt(50)), c);
  const h = v(ee);
  k(
    I(
      o,
      B(({ scrollTop: a }) => a),
      nt(),
      Ot(
        (a, x) => it(T) ? { direction: a.direction, prevScrollTop: x } : { direction: x < a.prevScrollTop ? ce : ee, prevScrollTop: x },
        { direction: ee, prevScrollTop: 0 }
      ),
      B((a) => a.direction)
    ),
    h
  ), k(I(o, zt(50), Bt(Ao)), h);
  const f = v(0);
  return k(
    I(
      p,
      P((a) => !a),
      Bt(0)
    ),
    f
  ), k(
    I(
      r,
      zt(100),
      $(p),
      P(([a, x]) => x),
      Ot(([a, x], [H]) => [x, H], [0, 0]),
      B(([a, x]) => x - a)
    ),
    f
  ), {
    atBottomState: w,
    atBottomStateChange: c,
    atBottomThreshold: m,
    atTopStateChange: d,
    atTopThreshold: S,
    isAtBottom: i,
    isAtTop: l,
    isScrolling: p,
    lastJumpDueToItemResize: R,
    scrollDirection: h,
    scrollVelocity: f
  };
}, rt(It)), ue = "top", ae = "bottom", gn = "none";
function In(t, e, n) {
  return typeof t == "number" ? n === ce && e === ue || n === ee && e === ae ? t : 0 : n === ce ? e === ue ? t.main : t.reverse : e === ae ? t.main : t.reverse;
}
function Sn(t, e) {
  return typeof t == "number" ? t : t[e] ?? 0;
}
const je = j(
  ([{ deviation: t, fixedHeaderHeight: e, headerHeight: n, scrollTop: o, viewportHeight: r }]) => {
    const s = U(), i = v(0), l = v(0), c = v(0), d = ht(
      I(
        at(
          W(o),
          W(r),
          W(n),
          W(s, ie),
          W(c),
          W(i),
          W(e),
          W(t),
          W(l)
        ),
        B(
          ([
            m,
            S,
            p,
            [T, w],
            R,
            h,
            f,
            a,
            x
          ]) => {
            const H = m - a, y = h + f, L = Math.max(p - H, 0);
            let u = gn;
            const g = Sn(x, ue), C = Sn(x, ae);
            return T -= a, T += p + f, w += p + f, w -= a, T > m + y - g && (u = ce), w < m - L + S + C && (u = ee), u !== gn ? [
              Math.max(H - p - In(R, ue, u) - g, 0),
              H - L - f + S + In(R, ae, u) + C
            ] : null;
          }
        ),
        P((m) => m !== null),
        nt(ie)
      ),
      [0, 0]
    );
    return {
      increaseViewportBy: l,
      // input
      listBoundary: s,
      overscan: c,
      topListHeight: i,
      // output
      visibleRange: d
    };
  },
  rt(It),
  { singleton: !0 }
);
function No(t, e, n) {
  if (ye(e)) {
    const o = Nn(t, e);
    return [
      { index: Rt(e.groupOffsetTree, o)[0], offset: 0, size: 0 },
      { data: n?.[0], index: o, offset: 0, size: 0 }
    ];
  }
  return [{ data: n?.[0], index: t, offset: 0, size: 0 }];
}
const Be = {
  bottom: 0,
  firstItemIndex: 0,
  items: [],
  offsetBottom: 0,
  offsetTop: 0,
  top: 0,
  topItems: [],
  topListHeight: 0,
  totalCount: 0
};
function xe(t, e, n, o, r, s) {
  const { lastIndex: i, lastOffset: l, lastSize: c } = r;
  let d = 0, m = 0;
  if (t.length > 0) {
    d = t[0].offset;
    const R = t[t.length - 1];
    m = R.offset + R.size;
  }
  const S = n - i, p = l + S * c + (S - 1) * o, T = d, w = p - m;
  return {
    bottom: m,
    firstItemIndex: s,
    items: xn(t, r, s),
    offsetBottom: w,
    offsetTop: d,
    top: T,
    topItems: xn(e, r, s),
    topListHeight: e.reduce((R, h) => h.size + R, 0),
    totalCount: n
  };
}
function jn(t, e, n, o, r, s) {
  let i = 0;
  if (n.groupIndices.length > 0)
    for (const m of n.groupIndices) {
      if (m - i >= t)
        break;
      i++;
    }
  const l = t + i, c = Ke(e, l), d = Array.from({ length: l }).map((m, S) => ({
    data: s[S + c],
    index: S + c,
    offset: 0,
    size: 0
  }));
  return xe(d, [], l, r, n, o);
}
function xn(t, e, n) {
  if (t.length === 0)
    return [];
  if (!ye(e))
    return t.map((d) => ({ ...d, index: d.index + n, originalIndex: d.index }));
  const o = t[0].index, r = t[t.length - 1].index, s = [], i = Yt(e.groupOffsetTree, o, r);
  let l, c = 0;
  for (const d of t) {
    (!l || l.end < d.index) && (l = i.shift(), c = e.groupIndices.indexOf(l.start));
    let m;
    d.index === l.start ? m = {
      index: c,
      type: "group"
    } : m = {
      groupIndex: c,
      index: d.index - (c + 1) + n
    }, s.push({
      ...m,
      data: d.data,
      offset: d.offset,
      originalIndex: d.index,
      size: d.size
    });
  }
  return s;
}
function Tn(t, e) {
  return t === void 0 ? 0 : typeof t == "number" ? t : t[e] ?? 0;
}
const Ut = j(
  ([
    { data: t, firstItemIndex: e, gap: n, sizes: o, totalCount: r },
    s,
    { listBoundary: i, topListHeight: l, visibleRange: c },
    { initialTopMostItemIndex: d, scrolledToInitialItem: m },
    { topListHeight: S },
    p,
    { didMount: T },
    { recalcInProgress: w }
  ]) => {
    const R = v([]), h = v(0), f = U(), a = v(0);
    k(s.topItemsIndexes, R);
    const x = ht(
      I(
        at(
          T,
          w,
          W(c, ie),
          W(r),
          W(o),
          W(d),
          m,
          W(R),
          W(e),
          W(n),
          W(a),
          t
        ),
        P(([u, g, , C, , , , , , , , z]) => {
          const O = z !== void 0 && z.length !== C;
          return u && !g && !O;
        }),
        B(
          ([
            ,
            ,
            [u, g],
            C,
            z,
            O,
            V,
            N,
            Z,
            F,
            mt,
            q
          ]) => {
            const Q = z, { offsetTree: gt, sizeTree: ut } = Q, St = it(h);
            if (C === 0)
              return { ...Be, totalCount: C };
            if (u === 0 && g === 0)
              return St === 0 ? { ...Be, totalCount: C } : jn(St, O, z, Z, F, q || []);
            if (J(ut))
              return St > 0 ? null : xe(
                No(Ke(O, C), Q, q),
                [],
                C,
                F,
                Q,
                Z
              );
            const Ft = [];
            if (N.length > 0) {
              const D = N[0], K = N[N.length - 1];
              let st = 0;
              for (const tt of Yt(ut, D, K)) {
                const X = tt.value, lt = Math.max(tt.start, D), xt = Math.min(tt.end, K);
                for (let ct = lt; ct <= xt; ct++)
                  Ft.push({ data: q?.[ct], index: ct, offset: st, size: X }), st += X;
              }
            }
            if (!V)
              return xe([], Ft, C, F, Q, Z);
            const pt = N.length > 0 ? N[N.length - 1] + 1 : 0, Kt = Oo(gt, u, g, pt);
            if (Kt.length === 0)
              return null;
            const Jt = C - 1, Et = we([], (D) => {
              for (const K of Kt) {
                const st = K.value;
                let tt = st.offset, X = K.start;
                const lt = st.size;
                if (st.offset < u) {
                  X += Math.floor((u - st.offset + F) / (lt + F));
                  const ct = X - K.start;
                  tt += ct * lt + ct * F;
                }
                X < pt && (tt += (pt - X) * lt, X = pt);
                const xt = Math.min(K.end, Jt);
                for (let ct = X; ct <= xt && !(tt >= g); ct++)
                  D.push({ data: q?.[ct], index: ct, offset: tt, size: lt }), tt += lt + F;
              }
            }), Qt = Tn(mt, ue), b = Tn(mt, ae);
            if (Et.length > 0 && (Qt > 0 || b > 0)) {
              const D = Et[0], K = Et[Et.length - 1];
              if (Qt > 0 && D.index > pt) {
                const st = Math.min(Qt, D.index - pt), tt = [];
                let X = D.offset;
                for (let lt = D.index - 1; lt >= D.index - st; lt--) {
                  const ct = Yt(ut, lt, lt)[0]?.value ?? D.size;
                  X -= ct + F, tt.unshift({ data: q?.[lt], index: lt, offset: X, size: ct });
                }
                Et.unshift(...tt);
              }
              if (b > 0 && K.index < Jt) {
                const st = Math.min(b, Jt - K.index);
                let tt = K.offset + K.size + F;
                for (let X = K.index + 1; X <= K.index + st; X++) {
                  const xt = Yt(ut, X, X)[0]?.value ?? K.size;
                  Et.push({ data: q?.[X], index: X, offset: tt, size: xt }), tt += xt + F;
                }
              }
            }
            return xe(Et, Ft, C, F, Q, Z);
          }
        ),
        //@ts-expect-error filter needs to be fixed
        P((u) => u !== null),
        nt()
      ),
      Be
    );
    k(
      I(
        t,
        P(Ae),
        B((u) => u?.length)
      ),
      r
    ), k(
      I(
        x,
        B((u) => u.topListHeight)
      ),
      S
    ), k(S, l), k(
      I(
        x,
        B((u) => [u.top, u.bottom])
      ),
      i
    ), k(
      I(
        x,
        B((u) => u.items)
      ),
      f
    );
    const H = vt(
      I(
        x,
        P(({ items: u }) => u.length > 0),
        $(r, t),
        P(([{ items: u }, g]) => u[u.length - 1].originalIndex === g - 1),
        B(([, u, g]) => [u - 1, g]),
        nt(ie),
        B(([u]) => u)
      )
    ), y = vt(
      I(
        x,
        zt(200),
        P(({ items: u, topItems: g }) => u.length > 0 && u[0].originalIndex === g.length),
        B(({ items: u }) => u[0].index),
        nt()
      )
    ), L = vt(
      I(
        x,
        P(({ items: u }) => u.length > 0),
        B(({ items: u }) => {
          let g = 0, C = u.length - 1;
          for (; u[g].type === "group" && g < C; )
            g++;
          for (; u[C].type === "group" && C > g; )
            C--;
          return {
            endIndex: u[C].index,
            startIndex: u[g].index
          };
        }),
        nt(Mn)
      )
    );
    return {
      endReached: H,
      initialItemCount: h,
      itemsRendered: f,
      listState: x,
      minOverscanItemCount: a,
      rangeChanged: L,
      startReached: y,
      topItemsIndexes: R,
      ...p
    };
  },
  rt(
    kt,
    $n,
    je,
    fe,
    de,
    me,
    At,
    De
  ),
  { singleton: !0 }
), qn = j(
  ([{ fixedFooterHeight: t, fixedHeaderHeight: e, footerHeight: n, headerHeight: o }, { listState: r }]) => {
    const s = U(), i = ht(
      I(
        at(n, t, o, e, r),
        B(([l, c, d, m, S]) => l + c + d + m + S.offsetBottom + S.bottom)
      ),
      0
    );
    return k(W(i), s), { totalListHeight: i, totalListHeightChanged: s };
  },
  rt(It, Ut),
  { singleton: !0 }
), Do = j(
  ([{ viewportHeight: t }, { totalListHeight: e }]) => {
    const n = v(!1), o = ht(
      I(
        at(n, t, e),
        P(([r]) => r),
        B(([, r, s]) => Math.max(0, r - s)),
        zt(0),
        nt()
      ),
      0
    );
    return { alignToBottom: n, paddingTopAddition: o };
  },
  rt(It, qn),
  { singleton: !0 }
), Yn = j(() => ({
  context: v(null)
})), $o = ({
  itemBottom: t,
  itemTop: e,
  locationParams: { align: n, behavior: o, ...r },
  viewportBottom: s,
  viewportTop: i
}) => e < i ? { ...r, align: n ?? "start", ...o !== void 0 ? { behavior: o } : {} } : t > s ? { ...r, align: n ?? "end", ...o !== void 0 ? { behavior: o } : {} } : null, Zn = j(
  ([
    { gap: t, sizes: e, totalCount: n },
    { fixedFooterHeight: o, fixedHeaderHeight: r, headerHeight: s, scrollingInProgress: i, scrollTop: l, viewportHeight: c },
    { scrollToIndex: d }
  ]) => {
    const m = U();
    return k(
      I(
        m,
        $(e, c, n, s, r, o, l),
        $(t),
        B(([[S, p, T, w, R, h, f, a], x]) => {
          const { calculateViewLocation: H = $o, done: y, ...L } = S, u = Dn(S, p, w - 1), g = le(u, p.offsetTree, x) + R + h, C = g + Rt(p.sizeTree, u)[1], z = a + h, O = a + T - f, V = H({
            itemBottom: C,
            itemTop: g,
            locationParams: L,
            viewportBottom: O,
            viewportTop: z
          });
          return V !== null ? y && yt(
            I(
              i,
              P((N) => !N),
              // skips the initial publish of false, and the cleanup call.
              // but if scrollingInProgress is true, we skip the initial publish.
              $t(it(i) ? 1 : 2)
            ),
            y
          ) : y?.(), V;
        }),
        P((S) => S !== null)
      ),
      d
    ), {
      scrollIntoView: m
    };
  },
  rt(kt, It, de, Ut, Gt),
  { singleton: !0 }
);
function vn(t) {
  return t === !1 ? !1 : t === "smooth" ? "smooth" : "auto";
}
const Uo = (t, e) => typeof t == "function" ? vn(t(e)) : e && vn(t), Ko = j(
  ([
    { listRefresh: t, totalCount: e, fixedItemSize: n, data: o },
    { atBottomState: r, isAtBottom: s },
    { scrollToIndex: i },
    { scrolledToInitialItem: l },
    { didMount: c, propsReady: d },
    { log: m },
    { scrollingInProgress: S },
    { context: p },
    { scrollIntoView: T }
  ]) => {
    const w = v(!1), R = U();
    let h = null;
    function f(y) {
      _(i, {
        align: "end",
        behavior: y,
        index: "LAST"
      });
    }
    Y(
      I(
        at(I(W(e), $t(1)), c),
        $(W(w), s, l, S),
        B(([[y, L], u, g, C, z]) => {
          let O = L && C, V = "auto";
          return O && (V = Uo(u, g || z), O = O && V !== !1), { followOutputBehavior: V, shouldFollow: O, totalCount: y };
        }),
        P(({ shouldFollow: y }) => y)
      ),
      ({ followOutputBehavior: y, totalCount: L }) => {
        h !== null && (h(), h = null), it(n) !== void 0 ? requestAnimationFrame(() => {
          it(m)("following output to ", { totalCount: L }, ft.DEBUG), f(y);
        }) : h = yt(t, () => {
          it(m)("following output to ", { totalCount: L }, ft.DEBUG), f(y), h = null;
        });
      }
    );
    function a(y) {
      const L = yt(r, (u) => {
        y && !u.atBottom && u.notAtBottomBecause === "SIZE_INCREASED" && h === null && (it(m)("scrolling to bottom due to increased size", {}, ft.DEBUG), f("auto"));
      });
      setTimeout(L, 100);
    }
    Y(
      I(
        at(W(w), e, d),
        P(([y, , L]) => y !== !1 && L),
        Ot(
          ({ value: y }, [, L]) => ({ refreshed: y === L, value: L }),
          { refreshed: !1, value: 0 }
        ),
        P(({ refreshed: y }) => y),
        $(w, e)
      ),
      ([, y]) => {
        it(l) && a(y !== !1);
      }
    ), Y(R, () => {
      a(it(w) !== !1);
    }), Y(at(W(w), r), ([y, L]) => {
      y !== !1 && !L.atBottom && L.notAtBottomBecause === "VIEWPORT_HEIGHT_DECREASING" && f("auto");
    });
    const x = v(null), H = U();
    return k(
      ze(
        I(
          W(o),
          B((y) => y?.length ?? 0)
        ),
        I(W(e))
      ),
      H
    ), Y(
      I(
        at(I(H, $t(1)), c),
        $(W(x), l, S, p),
        B(([[y, L], u, g, C, z]) => L && g && u?.({ context: z, totalCount: y, scrollingInProgress: C })),
        P((y) => !!y),
        zt(0)
      ),
      (y) => {
        h !== null && (h(), h = null), it(n) !== void 0 ? requestAnimationFrame(() => {
          it(m)("scrolling into view", {}), _(T, y);
        }) : h = yt(t, () => {
          it(m)("scrolling into view", {}), _(T, y), h = null;
        });
      }
    ), { autoscrollToBottom: R, followOutput: w, scrollIntoViewOnChange: x };
  },
  rt(
    kt,
    me,
    de,
    fe,
    At,
    Gt,
    It,
    Yn,
    Zn
  )
), jo = j(
  ([{ data: t, firstItemIndex: e, gap: n, sizes: o }, { initialTopMostItemIndex: r }, { initialItemCount: s, listState: i }, { didMount: l }]) => (k(
    I(
      l,
      $(s),
      P(([, c]) => c !== 0),
      $(r, o, e, n, t),
      B(([[, c], d, m, S, p, T = []]) => jn(c, d, m, S, p, T))
    ),
    i
  ), {}),
  rt(kt, fe, Ut, At),
  { singleton: !0 }
), qo = j(
  ([{ didMount: t }, { scrollTo: e }, { listState: n }]) => {
    const o = v(0);
    return Y(
      I(
        t,
        $(o),
        P(([, r]) => r !== 0),
        B(([, r]) => ({ top: r }))
      ),
      (r) => {
        yt(
          I(
            n,
            $t(1),
            P((s) => s.items.length > 1)
          ),
          () => {
            requestAnimationFrame(() => {
              _(e, r);
            });
          }
        );
      }
    ), {
      initialScrollTop: o
    };
  },
  rt(At, It, Ut),
  { singleton: !0 }
), Xn = j(
  ([{ scrollVelocity: t }]) => {
    const e = v(!1), n = U(), o = v(!1);
    return k(
      I(
        t,
        $(o, e, n),
        P(([r, s]) => s !== !1 && s !== void 0),
        B(([r, s, i, l]) => {
          const { enter: c, exit: d } = s;
          if (i) {
            if (d(r, l))
              return !1;
          } else if (c(r, l))
            return !0;
          return i;
        }),
        nt()
      ),
      e
    ), Y(
      I(at(e, t, n), $(o)),
      ([[r, s, i], l]) => {
        r && l !== !1 && l !== void 0 && l.change && l.change(s, i);
      }
    ), { isSeeking: e, scrollSeekConfiguration: o, scrollSeekRangeChanged: n, scrollVelocity: t };
  },
  rt(me),
  { singleton: !0 }
), qe = j(([{ scrollContainerState: t, scrollTo: e }]) => {
  const n = U(), o = U(), r = U(), s = v(!1), i = v(void 0);
  return k(
    I(
      at(n, o),
      B(([{ scrollTop: l, viewportHeight: c }, { offsetTop: d, listHeight: m }]) => ({
        scrollHeight: m,
        scrollTop: Math.max(0, l - d),
        viewportHeight: c
      }))
    ),
    t
  ), k(
    I(
      e,
      $(o),
      B(([l, { offsetTop: c }]) => ({
        ...l,
        top: l.top + c
      }))
    ),
    r
  ), {
    customScrollParent: i,
    // config
    useWindowScroll: s,
    // input
    windowScrollContainerState: n,
    // signals
    windowScrollTo: r,
    windowViewportRect: o
  };
}, rt(It)), Yo = j(
  ([
    { sizeRanges: t, sizes: e },
    { headerHeight: n, scrollTop: o },
    { initialTopMostItemIndex: r },
    { didMount: s },
    { useWindowScroll: i, windowScrollContainerState: l, windowViewportRect: c }
  ]) => {
    const d = U(), m = v(void 0), S = v(null), p = v(null);
    return k(l, S), k(c, p), Y(
      I(
        d,
        $(e, o, i, S, p, n)
      ),
      ([T, w, R, h, f, a, x]) => {
        const H = ko(w.sizeTree);
        h && f !== null && a !== null && (R = f.scrollTop - a.offsetTop), R -= x, T({ ranges: H, scrollTop: R });
      }
    ), k(I(m, P(Ae), B(Zo)), r), k(
      I(
        s,
        $(m),
        P(([, T]) => T !== void 0),
        nt(),
        B(([, T]) => T.ranges)
      ),
      t
    ), {
      getState: d,
      restoreStateFrom: m
    };
  },
  rt(kt, It, fe, At, qe)
);
function Zo(t) {
  return { align: "start", index: 0, offset: t.scrollTop };
}
const Xo = j(([{ topItemsIndexes: t }]) => {
  const e = v(0);
  return k(
    I(
      e,
      P((n) => n >= 0),
      B((n) => Array.from({ length: n }).map((o, r) => r))
    ),
    t
  ), { topItemCount: e };
}, rt(Ut));
function Jn(t) {
  let e = !1, n;
  return (() => (e || (e = !0, n = t()), n));
}
const Jo = Jn(() => /iP(ad|od|hone)/i.test(navigator.userAgent) && /WebKit/i.test(navigator.userAgent)), Qo = j(
  ([
    { deviation: t, scrollBy: e, scrollingInProgress: n, scrollTop: o },
    { isAtBottom: r, isScrolling: s, lastJumpDueToItemResize: i, scrollDirection: l },
    { listState: c },
    { beforeUnshiftWith: d, gap: m, shiftWithOffset: S, sizes: p },
    { log: T },
    { recalcInProgress: w }
  ]) => {
    const R = vt(
      I(
        c,
        $(i),
        Ot(
          ([, f, a, x], [{ bottom: H, items: y, offsetBottom: L, totalCount: u }, g]) => {
            const C = H + L;
            let z = 0;
            return a === u && f.length > 0 && y.length > 0 && (y[0].originalIndex === 0 && f[0].originalIndex === 0 || (z = C - x, z !== 0 && (z += g))), [z, y, u, C];
          },
          [0, [], 0, 0]
        ),
        P(([f]) => f !== 0),
        $(o, l, n, r, T, w),
        P(([, f, a, x, , , H]) => !H && !x && f !== 0 && a === ce),
        B(([[f], , , , , a]) => (a("Upward scrolling compensation", { amount: f }, ft.DEBUG), f))
      )
    );
    function h(f) {
      f > 0 ? (_(e, { behavior: "auto", top: -f }), _(t, 0)) : (_(t, 0), _(e, { behavior: "auto", top: -f }));
    }
    return Y(I(R, $(t, s)), ([f, a, x]) => {
      x && Jo() ? _(t, a - f) : h(-f);
    }), Y(
      I(
        at(ht(s, !1), t, w),
        P(([f, a, x]) => !f && !x && a !== 0),
        B(([f, a]) => a),
        zt(1)
      ),
      h
    ), k(
      I(
        S,
        B((f) => ({ top: -f }))
      ),
      e
    ), Y(
      I(
        d,
        $(p, m),
        B(([f, { groupIndices: a, lastSize: x, sizeTree: H }, y]) => {
          function L(O) {
            return O * (x + y);
          }
          if (a.length === 0)
            return L(f);
          let u = 0;
          const g = se(H, 0);
          let C = 0, z = 0;
          for (; C < f; ) {
            C++, u += g;
            let O = a.length === z + 1 ? 1 / 0 : a[z + 1] - a[z] - 1;
            C + O > f && (u -= g, O = f - C + 1), C += O, u += L(O), z++;
          }
          return u;
        })
      ),
      (f) => {
        _(t, f), requestAnimationFrame(() => {
          _(e, { top: f }), requestAnimationFrame(() => {
            _(t, 0), _(w, !1);
          });
        });
      }
    ), { deviation: t };
  },
  rt(It, me, Ut, kt, Gt, De)
), tr = j(
  ([
    t,
    e,
    n,
    o,
    r,
    s,
    i,
    l,
    c,
    d,
    m
  ]) => ({
    ...t,
    ...e,
    ...n,
    ...o,
    ...r,
    ...s,
    ...i,
    ...l,
    ...c,
    ...d,
    ...m
  }),
  rt(
    je,
    jo,
    At,
    Xn,
    qn,
    qo,
    Do,
    qe,
    Zn,
    Gt,
    Yn
  )
), Qn = j(
  ([
    {
      data: t,
      defaultItemSize: e,
      firstItemIndex: n,
      fixedItemSize: o,
      fixedGroupSize: r,
      gap: s,
      groupIndices: i,
      heightEstimates: l,
      itemSize: c,
      sizeRanges: d,
      sizes: m,
      statefulTotalCount: S,
      totalCount: p,
      trackItemSizes: T
    },
    { initialItemFinalLocationReached: w, initialTopMostItemIndex: R, scrolledToInitialItem: h },
    f,
    a,
    x,
    H,
    { scrollToIndex: y },
    L,
    { topItemCount: u },
    { groupCounts: g },
    C
  ]) => {
    const { listState: z, minOverscanItemCount: O, topItemsIndexes: V, rangeChanged: N, ...Z } = H;
    return k(N, C.scrollSeekRangeChanged), k(
      I(
        C.windowViewportRect,
        B((F) => F.visibleHeight)
      ),
      f.viewportHeight
    ), {
      data: t,
      defaultItemHeight: e,
      firstItemIndex: n,
      fixedItemHeight: o,
      fixedGroupHeight: r,
      gap: s,
      groupCounts: g,
      heightEstimates: l,
      initialItemFinalLocationReached: w,
      initialTopMostItemIndex: R,
      scrolledToInitialItem: h,
      sizeRanges: d,
      topItemCount: u,
      topItemsIndexes: V,
      // input
      totalCount: p,
      ...x,
      groupIndices: i,
      itemSize: c,
      listState: z,
      minOverscanItemCount: O,
      scrollToIndex: y,
      // output
      statefulTotalCount: S,
      trackItemSizes: T,
      // exported from stateFlagsSystem
      rangeChanged: N,
      ...Z,
      // the bag of IO from featureGroup1System
      ...C,
      ...f,
      sizes: m,
      ...a
    };
  },
  rt(
    kt,
    fe,
    It,
    Yo,
    Ko,
    Ut,
    de,
    Qo,
    Xo,
    $n,
    tr
  )
);
function er(t, e) {
  const n = {}, o = {};
  let r = 0;
  const s = t.length;
  for (; r < s; )
    o[t[r]] = 1, r += 1;
  for (const i in e)
    Object.hasOwn(o, i) || (n[i] = e[i]);
  return n;
}
const ge = typeof document < "u" ? E.useLayoutEffect : E.useEffect;
function Ye(t, e, n) {
  const o = Object.keys(e.required || {}), r = Object.keys(e.optional || {}), s = Object.keys(e.methods || {}), i = Object.keys(e.events || {}), l = E.createContext({});
  function c(f, a) {
    f.propsReady !== void 0 && _(f.propsReady, !1);
    for (const x of o) {
      const H = f[e.required[x]];
      _(H, a[x]);
    }
    for (const x of r)
      if (x in a) {
        const H = f[e.optional[x]];
        _(H, a[x]);
      }
    f.propsReady !== void 0 && _(f.propsReady, !0);
  }
  function d(f) {
    return s.reduce((a, x) => (a[x] = (H) => {
      const y = f[e.methods[x]];
      _(y, H);
    }, a), {});
  }
  function m(f) {
    return i.reduce((a, x) => (a[x] = So(f[e.events[x]]), a), {});
  }
  const S = E.forwardRef(function(a, x) {
    const { children: H, ...y } = a, [L] = E.useState(() => we(To(t), (C) => {
      c(C, y);
    })), [u] = E.useState(dn(m, L));
    ge(() => {
      for (const C of i)
        C in y && Y(u[C], y[C]);
      return () => {
        Object.values(u).map(Me);
      };
    }, [y, u, L]), ge(() => {
      c(L, y);
    }), E.useImperativeHandle(x, an(d(L)));
    const g = n;
    return /* @__PURE__ */ G(l.Provider, { value: L, children: n !== void 0 ? /* @__PURE__ */ G(g, { ...er([...o, ...r, ...i], y), children: H }) : H });
  }), p = (f) => {
    const a = E.useContext(l);
    return E.useCallback(
      (x) => {
        _(a[f], x);
      },
      [a, f]
    );
  }, T = (f) => {
    const x = E.useContext(l)[f], H = E.useCallback(
      (y) => Y(x, y),
      [x]
    );
    return E.useSyncExternalStore(
      H,
      () => it(x),
      () => it(x)
    );
  }, w = (f) => {
    const x = E.useContext(l)[f], [H, y] = E.useState(dn(it, x));
    return ge(
      () => Y(x, (L) => {
        L !== H && y(an(L));
      }),
      [x, H]
    ), H;
  }, R = parseInt(E.version) >= 18 ? T : w;
  return {
    Component: S,
    useEmitter: (f, a) => {
      const H = E.useContext(l)[f];
      ge(() => Y(H, a), [a, H]);
    },
    useEmitterValue: R,
    usePublisher: p
  };
}
const be = E.createContext(void 0), to = E.createContext(void 0), Oe = "-webkit-sticky", Cn = "sticky", Ze = Jn(() => {
  if (typeof document > "u")
    return Cn;
  const t = document.createElement("div");
  return t.style.position = Oe, t.style.position === Oe ? Oe : Cn;
}), eo = typeof document < "u" ? E.useLayoutEffect : E.useEffect;
function Le(t) {
  return "self" in t;
}
function nr(t) {
  return "body" in t;
}
function no(t, e, n, o = Xt, r, s) {
  const i = E.useRef(null), l = E.useRef(null), c = E.useRef(null), d = E.useCallback(
    (p) => {
      let T, w, R;
      const h = p.target;
      if (nr(h) || Le(h)) {
        const a = Le(h) ? h : h.defaultView;
        R = s === !0 ? a.scrollX : a.scrollY, T = s === !0 ? a.document.documentElement.scrollWidth : a.document.documentElement.scrollHeight, w = s === !0 ? a.innerWidth : a.innerHeight;
      } else
        R = s === !0 ? h.scrollLeft : h.scrollTop, T = s === !0 ? h.scrollWidth : h.scrollHeight, w = s === !0 ? h.offsetWidth : h.offsetHeight;
      const f = () => {
        t({
          scrollHeight: T,
          scrollTop: Math.max(R, 0),
          viewportHeight: w
        });
      };
      p.suppressFlushSync === !0 ? f() : fo.flushSync(f), l.current !== null && (R === l.current || R <= 0 || R === T - w) && (l.current = null, e(!0), c.current && (clearTimeout(c.current), c.current = null));
    },
    [t, e, s]
  );
  E.useEffect(() => {
    const p = r || i.current;
    return o(r || i.current), d({ suppressFlushSync: !0, target: p }), p.addEventListener("scroll", d, { passive: !0 }), () => {
      o(null), p.removeEventListener("scroll", d);
    };
  }, [i, d, n, o, r]);
  function m(p) {
    const T = i.current;
    if (!T || (s === !0 ? "offsetWidth" in T && T.offsetWidth === 0 : "offsetHeight" in T && T.offsetHeight === 0))
      return;
    const w = p.behavior === "smooth";
    let R, h, f;
    Le(T) ? (h = Math.max(
      Ht(T.document.documentElement, s === !0 ? "width" : "height"),
      s === !0 ? T.document.documentElement.scrollWidth : T.document.documentElement.scrollHeight
    ), R = s === !0 ? T.innerWidth : T.innerHeight, f = s === !0 ? window.scrollX : window.scrollY) : (h = T[s === !0 ? "scrollWidth" : "scrollHeight"], R = Ht(T, s === !0 ? "width" : "height"), f = T[s === !0 ? "scrollLeft" : "scrollTop"]);
    const a = h - R;
    if (p.top = Math.ceil(Math.max(Math.min(a, p.top), 0)), Kn(R, h) || p.top === f) {
      t({ scrollHeight: h, scrollTop: f, viewportHeight: R }), w && e(!0);
      return;
    }
    w ? (l.current = p.top, c.current && clearTimeout(c.current), c.current = setTimeout(() => {
      c.current = null, l.current = null, e(!0);
    }, 1e3)) : l.current = null, s === !0 && (p = { ...p.behavior !== void 0 ? { behavior: p.behavior } : {}, left: p.top }), T.scrollTo(p);
  }
  function S(p) {
    s === !0 && (p = {
      ...p.behavior !== void 0 ? { behavior: p.behavior } : {},
      ...p.top !== void 0 ? { left: p.top } : {}
    }), i.current.scrollBy(p);
  }
  return { scrollByCallback: S, scrollerRef: i, scrollToCallback: m };
}
function Xe(t) {
  return t;
}
const or = /* @__PURE__ */ j(() => {
  const t = v((l) => `Item ${l}`), e = v((l) => `Group ${l}`), n = v({}), o = v(Xe), r = v("div"), s = v(Xt), i = (l, c = null) => ht(
    I(
      n,
      B((d) => d[l]),
      nt()
    ),
    c
  );
  return {
    components: n,
    computeItemKey: o,
    EmptyPlaceholder: i("EmptyPlaceholder"),
    FooterComponent: i("Footer"),
    GroupComponent: i("Group", "div"),
    groupContent: e,
    HeaderComponent: i("Header"),
    HeaderFooterTag: r,
    ItemComponent: i("Item", "div"),
    itemContent: t,
    ListComponent: i("List", "div"),
    ScrollerComponent: i("Scroller", "div"),
    scrollerRef: s,
    ScrollSeekPlaceholder: i("ScrollSeekPlaceholder"),
    TopItemListComponent: i("TopItemList")
  };
}), rr = /* @__PURE__ */ j(
  ([t, e]) => ({ ...t, ...e }),
  rt(Qn, or)
), sr = ({ height: t }) => /* @__PURE__ */ G("div", { style: { height: t } }), ir = { overflowAnchor: "none", position: Ze(), zIndex: 1 }, oo = { overflowAnchor: "none" }, lr = { ...oo, display: "inline-block", height: "100%" }, wn = /* @__PURE__ */ E.memo(function({ showTopList: e = !1 }) {
  const n = A("listState"), o = Ct("sizeRanges"), r = A("useWindowScroll"), s = A("customScrollParent"), i = Ct("windowScrollContainerState"), l = Ct("scrollContainerState"), c = s || r ? i : l, d = A("itemContent"), m = A("context"), S = A("groupContent"), p = A("trackItemSizes"), T = A("itemSize"), w = A("log"), R = Ct("gap"), h = A("horizontalDirection"), { callbackRef: f } = Fn(
    o,
    T,
    p,
    e ? Xt : c,
    w,
    R,
    s,
    h,
    A("skipAnimationFrameInResizeObserver")
  ), [a, x] = E.useState(0);
  en("deviation", (F) => {
    a !== F && x(F);
  });
  const H = A("EmptyPlaceholder"), y = A("ScrollSeekPlaceholder") ?? sr, L = A("ListComponent"), u = A("ItemComponent"), g = A("GroupComponent"), C = A("computeItemKey"), z = A("isSeeking"), O = A("groupIndices").length > 0, V = A("alignToBottom"), N = A("initialItemFinalLocationReached"), Z = e ? {} : {
    boxSizing: "border-box",
    ...h ? {
      display: "inline-block",
      height: "100%",
      marginLeft: a !== 0 ? a : V ? "auto" : 0,
      paddingLeft: n.offsetTop,
      paddingRight: n.offsetBottom,
      whiteSpace: "nowrap"
    } : {
      marginTop: a !== 0 ? a : V ? "auto" : 0,
      paddingBottom: n.offsetBottom,
      paddingTop: n.offsetTop
    },
    ...N ? {} : { visibility: "hidden" }
  };
  return !e && n.totalCount === 0 && H !== null && H !== void 0 ? /* @__PURE__ */ G(H, { ...ot(H, m) }) : /* @__PURE__ */ G(
    L,
    {
      ...ot(L, m),
      "data-testid": e ? "virtuoso-top-item-list" : "virtuoso-item-list",
      ref: f,
      style: Z,
      children: (e ? n.topItems : n.items).map((F) => {
        const mt = F.originalIndex, q = C(mt + n.firstItemIndex, F.data, m);
        return z ? /* @__PURE__ */ Nt(
          y,
          {
            ...ot(y, m),
            height: F.size,
            index: F.index,
            key: q,
            type: F.type || "item",
            ...F.type === "group" ? {} : { groupIndex: F.groupIndex }
          }
        ) : F.type === "group" ? /* @__PURE__ */ Nt(
          g,
          {
            ...ot(g, m),
            "data-index": mt,
            "data-item-index": F.index,
            "data-known-size": F.size,
            key: q,
            style: ir
          },
          S(F.index, m)
        ) : /* @__PURE__ */ Nt(
          u,
          {
            ...ot(u, m),
            ...ro(u, F.data),
            "data-index": mt,
            "data-item-group-index": F.groupIndex,
            "data-item-index": F.index,
            "data-known-size": F.size,
            key: q,
            style: h ? lr : oo
          },
          O ? d(F.index, F.groupIndex, F.data, m) : d(F.index, F.data, m)
        );
      })
    }
  );
}), cr = {
  height: "100%",
  outline: "none",
  overflowY: "auto",
  position: "relative",
  WebkitOverflowScrolling: "touch"
}, ur = {
  outline: "none",
  overflowX: "auto",
  position: "relative"
}, Re = (t) => ({
  height: "100%",
  position: "absolute",
  top: 0,
  width: "100%",
  ...t ? { display: "flex", flexDirection: "column" } : void 0
}), Je = (t, e, n = 0) => ({
  ...Re(t),
  position: e ? "relative" : "absolute",
  top: e ? -n : 0
}), ar = {
  position: Ze(),
  top: 0,
  width: "100%",
  zIndex: 1
};
function ot(t, e) {
  if (typeof t != "string")
    return { context: e };
}
function ro(t, e) {
  return { item: typeof t == "string" ? void 0 : e };
}
const dr = /* @__PURE__ */ E.memo(function() {
  const e = A("HeaderComponent"), n = Ct("headerHeight"), o = A("HeaderFooterTag"), r = Lt(
    E.useMemo(
      () => (i) => {
        n(Ht(i, "height"));
      },
      [n]
    ),
    !0,
    A("skipAnimationFrameInResizeObserver")
  ), s = A("context");
  return e != null ? /* @__PURE__ */ G(o, { ref: r, children: /* @__PURE__ */ G(e, { ...ot(e, s) }) }) : null;
}), fr = /* @__PURE__ */ E.memo(function() {
  const e = A("FooterComponent"), n = Ct("footerHeight"), o = A("HeaderFooterTag"), r = Lt(
    E.useMemo(
      () => (i) => {
        n(Ht(i, "height"));
      },
      [n]
    ),
    !0,
    A("skipAnimationFrameInResizeObserver")
  ), s = A("context");
  return e != null ? /* @__PURE__ */ G(o, { ref: r, children: /* @__PURE__ */ G(e, { ...ot(e, s) }) }) : null;
});
function Qe({ useEmitter: t, useEmitterValue: e, usePublisher: n }) {
  return E.memo(function({ children: s, style: i, context: l, ...c }) {
    const d = n("scrollContainerState"), m = e("ScrollerComponent"), S = n("smoothScrollTargetReached"), p = e("scrollerRef"), T = e("horizontalDirection") || !1, { scrollByCallback: w, scrollerRef: R, scrollToCallback: h } = no(
      d,
      S,
      m,
      p,
      void 0,
      T
    );
    return t("scrollTo", h), t("scrollBy", w), /* @__PURE__ */ G(
      m,
      {
        "data-testid": "virtuoso-scroller",
        "data-virtuoso-scroller": !0,
        ref: R,
        style: { ...T ? ur : cr, ...i },
        tabIndex: 0,
        ...c,
        ...ot(m, l),
        children: s
      }
    );
  });
}
function tn({ useEmitter: t, useEmitterValue: e, usePublisher: n }) {
  return E.memo(function({ children: s, style: i, context: l, ...c }) {
    const d = n("windowScrollContainerState"), m = e("ScrollerComponent"), S = n("smoothScrollTargetReached"), p = e("totalListHeight"), T = e("deviation"), w = e("customScrollParent"), R = E.useRef(null), h = e("scrollerRef"), { scrollByCallback: f, scrollerRef: a, scrollToCallback: x } = no(
      d,
      S,
      m,
      h,
      w
    );
    return eo(() => (a.current = w || R.current?.ownerDocument.defaultView, () => {
      a.current = null;
    }), [a, w]), t("windowScrollTo", x), t("scrollBy", f), /* @__PURE__ */ G(
      m,
      {
        ref: R,
        "data-virtuoso-scroller": !0,
        style: { position: "relative", ...i, ...p !== 0 ? { height: p + T } : void 0 },
        ...c,
        ...ot(m, l),
        children: s
      }
    );
  });
}
const mr = ({ children: t }) => {
  const e = E.useContext(be), n = Ct("viewportHeight"), o = Ct("fixedItemHeight"), r = A("alignToBottom"), s = A("horizontalDirection"), i = E.useMemo(
    () => oe(n, (c) => Ht(c, s ? "width" : "height")),
    [n, s]
  ), l = Lt(i, !0, A("skipAnimationFrameInResizeObserver"));
  return E.useEffect(() => {
    e && (n(e.viewportHeight), o(e.itemHeight));
  }, [e, n, o]), /* @__PURE__ */ G("div", { "data-viewport-type": "element", ref: l, style: Re(r), children: t });
}, pr = ({ children: t }) => {
  const e = E.useContext(be), n = Ct("windowViewportRect"), o = Ct("fixedItemHeight"), r = A("customScrollParent"), s = A("useWindowScroll"), i = A("topListHeight"), l = Ne(
    n,
    r,
    A("skipAnimationFrameInResizeObserver")
  ), c = A("alignToBottom");
  return E.useEffect(() => {
    e && (o(e.itemHeight), n({ listHeight: 0, offsetTop: 0, visibleHeight: e.viewportHeight, visibleWidth: 100 }));
  }, [e, n, o]), /* @__PURE__ */ G("div", { "data-viewport-type": "window", ref: l, style: Je(c, s, i), children: t });
}, hr = ({ children: t }) => {
  const e = A("TopItemListComponent") ?? "div", n = A("headerHeight"), o = { ...ar, marginTop: `${n}px` }, r = A("context");
  return /* @__PURE__ */ G(e, { style: o, ...ot(e, r), children: t });
}, gr = /* @__PURE__ */ E.memo(function(e) {
  const n = A("useWindowScroll"), o = A("topItemsIndexes").length > 0, r = A("customScrollParent"), s = A("context");
  return /* @__PURE__ */ Dt(r || n ? Sr : Ir, { ...e, context: s, children: [
    o && /* @__PURE__ */ G(hr, { children: /* @__PURE__ */ G(wn, { showTopList: !0 }) }),
    /* @__PURE__ */ Dt(r || n ? pr : mr, { children: [
      /* @__PURE__ */ G(dr, {}),
      /* @__PURE__ */ G(wn, {}),
      /* @__PURE__ */ G(fr, {})
    ] })
  ] });
}), {
  Component: so,
  useEmitter: en,
  useEmitterValue: A,
  usePublisher: Ct
} = /* @__PURE__ */ Ye(
  rr,
  {
    optional: {
      restoreStateFrom: "restoreStateFrom",
      context: "context",
      followOutput: "followOutput",
      scrollIntoViewOnChange: "scrollIntoViewOnChange",
      itemContent: "itemContent",
      groupContent: "groupContent",
      overscan: "overscan",
      increaseViewportBy: "increaseViewportBy",
      minOverscanItemCount: "minOverscanItemCount",
      totalCount: "totalCount",
      groupCounts: "groupCounts",
      topItemCount: "topItemCount",
      firstItemIndex: "firstItemIndex",
      initialTopMostItemIndex: "initialTopMostItemIndex",
      components: "components",
      atBottomThreshold: "atBottomThreshold",
      atTopThreshold: "atTopThreshold",
      computeItemKey: "computeItemKey",
      defaultItemHeight: "defaultItemHeight",
      fixedGroupHeight: "fixedGroupHeight",
      // Must be set above 'fixedItemHeight'
      fixedItemHeight: "fixedItemHeight",
      heightEstimates: "heightEstimates",
      itemSize: "itemSize",
      scrollSeekConfiguration: "scrollSeekConfiguration",
      headerFooterTag: "HeaderFooterTag",
      data: "data",
      initialItemCount: "initialItemCount",
      initialScrollTop: "initialScrollTop",
      alignToBottom: "alignToBottom",
      useWindowScroll: "useWindowScroll",
      customScrollParent: "customScrollParent",
      scrollerRef: "scrollerRef",
      logLevel: "logLevel",
      horizontalDirection: "horizontalDirection",
      skipAnimationFrameInResizeObserver: "skipAnimationFrameInResizeObserver"
    },
    methods: {
      scrollToIndex: "scrollToIndex",
      scrollIntoView: "scrollIntoView",
      scrollTo: "scrollTo",
      scrollBy: "scrollBy",
      autoscrollToBottom: "autoscrollToBottom",
      getState: "getState"
    },
    events: {
      isScrolling: "isScrolling",
      endReached: "endReached",
      startReached: "startReached",
      rangeChanged: "rangeChanged",
      atBottomStateChange: "atBottomStateChange",
      atTopStateChange: "atTopStateChange",
      totalListHeightChanged: "totalListHeightChanged",
      itemsRendered: "itemsRendered",
      groupIndices: "groupIndices"
    }
  },
  gr
), Ir = /* @__PURE__ */ Qe({ useEmitter: en, useEmitterValue: A, usePublisher: Ct }), Sr = /* @__PURE__ */ tn({ useEmitter: en, useEmitterValue: A, usePublisher: Ct }), Yr = so, Zr = so, xr = /* @__PURE__ */ j(() => {
  const t = v((d) => /* @__PURE__ */ Dt("td", { children: [
    "Item $",
    d
  ] })), e = v(null), n = v((d) => /* @__PURE__ */ Dt("td", { colSpan: 1e3, children: [
    "Group ",
    d
  ] })), o = v(null), r = v(null), s = v({}), i = v(Xe), l = v(Xt), c = (d, m = null) => ht(
    I(
      s,
      B((S) => S[d]),
      nt()
    ),
    m
  );
  return {
    components: s,
    computeItemKey: i,
    context: e,
    EmptyPlaceholder: c("EmptyPlaceholder"),
    FillerRow: c("FillerRow"),
    fixedFooterContent: r,
    fixedHeaderContent: o,
    itemContent: t,
    groupContent: n,
    ScrollerComponent: c("Scroller", "div"),
    scrollerRef: l,
    ScrollSeekPlaceholder: c("ScrollSeekPlaceholder"),
    TableBodyComponent: c("TableBody", "tbody"),
    TableComponent: c("Table", "table"),
    TableFooterComponent: c("TableFoot", "tfoot"),
    TableHeadComponent: c("TableHead", "thead"),
    TableRowComponent: c("TableRow", "tr"),
    GroupComponent: c("Group", "tr")
  };
}), Tr = /* @__PURE__ */ j(
  ([t, e]) => ({ ...t, ...e }),
  rt(Qn, xr)
), vr = ({ height: t }) => /* @__PURE__ */ G("tr", { children: /* @__PURE__ */ G("td", { style: { height: t } }) }), Cr = ({ height: t }) => /* @__PURE__ */ G("tr", { children: /* @__PURE__ */ G("td", { style: { border: 0, height: t, padding: 0 } }) }), wr = { overflowAnchor: "none" }, yn = { position: Ze(), zIndex: 2, overflowAnchor: "none" }, bn = /* @__PURE__ */ E.memo(function({ showTopList: e = !1 }) {
  const n = M("listState"), o = M("computeItemKey"), r = M("firstItemIndex"), s = M("context"), i = M("isSeeking"), l = M("fixedHeaderHeight"), c = M("groupIndices").length > 0, d = M("itemContent"), m = M("groupContent"), S = M("ScrollSeekPlaceholder") ?? vr, p = M("GroupComponent"), T = M("TableRowComponent"), w = (e ? n.topItems : []).reduce((h, f, a) => (a === 0 ? h.push(f.size) : h.push(h[a - 1] + f.size), h), []);
  return (e ? n.topItems : n.items).map((h) => {
    const f = h.originalIndex, a = o(f + r, h.data, s), x = e ? f === 0 ? 0 : w[f - 1] : 0;
    return i ? /* @__PURE__ */ Nt(
      S,
      {
        ...ot(S, s),
        height: h.size,
        index: h.index,
        key: a,
        type: h.type || "item"
      }
    ) : h.type === "group" ? /* @__PURE__ */ Nt(
      p,
      {
        ...ot(p, s),
        "data-index": f,
        "data-item-index": h.index,
        "data-known-size": h.size,
        key: a,
        style: {
          ...yn,
          top: l
        }
      },
      m(h.index, s)
    ) : /* @__PURE__ */ Nt(
      T,
      {
        ...ot(T, s),
        ...ro(T, h.data),
        "data-index": f,
        "data-item-index": h.index,
        "data-known-size": h.size,
        "data-item-group-index": h.groupIndex,
        key: a,
        style: e ? { ...yn, top: l + x } : wr
      },
      c ? d(h.index, h.groupIndex, h.data, s) : d(h.index, h.data, s)
    );
  });
}), yr = /* @__PURE__ */ E.memo(function() {
  const e = M("listState"), n = M("topItemsIndexes").length > 0, o = bt("sizeRanges"), r = M("useWindowScroll"), s = M("customScrollParent"), i = bt("windowScrollContainerState"), l = bt("scrollContainerState"), c = s || r ? i : l, d = M("trackItemSizes"), m = M("itemSize"), S = M("log"), { callbackRef: p, ref: T } = Fn(
    o,
    m,
    d,
    c,
    S,
    void 0,
    s,
    !1,
    M("skipAnimationFrameInResizeObserver")
  ), [w, R] = E.useState(0);
  nn("deviation", (O) => {
    w !== O && (T.current.style.marginTop = `${O}px`, R(O));
  });
  const h = M("EmptyPlaceholder"), f = M("FillerRow") ?? Cr, a = M("TableBodyComponent"), x = M("paddingTopAddition"), H = M("statefulTotalCount"), y = M("context");
  if (H === 0 && h !== null && h !== void 0)
    return /* @__PURE__ */ G(h, { ...ot(h, y) });
  const L = (n ? e.topItems : []).reduce((O, V) => O + V.size, 0), u = e.offsetTop + x + w - L, g = e.offsetBottom, C = u > 0 ? /* @__PURE__ */ G(f, { context: y, height: u }, "padding-top") : null, z = g > 0 ? /* @__PURE__ */ G(f, { context: y, height: g }, "padding-bottom") : null;
  return /* @__PURE__ */ Dt(a, { "data-testid": "virtuoso-item-list", ref: p, ...ot(a, y), children: [
    C,
    n && /* @__PURE__ */ G(bn, { showTopList: !0 }),
    /* @__PURE__ */ G(bn, {}),
    z
  ] });
}), br = ({ children: t }) => {
  const e = E.useContext(be), n = bt("viewportHeight"), o = bt("fixedItemHeight"), r = Lt(
    E.useMemo(() => oe(n, (s) => Ht(s, "height")), [n]),
    !0,
    M("skipAnimationFrameInResizeObserver")
  );
  return E.useEffect(() => {
    e && (n(e.viewportHeight), o(e.itemHeight));
  }, [e, n, o]), /* @__PURE__ */ G("div", { "data-viewport-type": "element", ref: r, style: Re(!1), children: t });
}, Rr = ({ children: t }) => {
  const e = E.useContext(be), n = bt("windowViewportRect"), o = bt("fixedItemHeight"), r = M("customScrollParent"), s = M("useWindowScroll"), i = Ne(
    n,
    r,
    M("skipAnimationFrameInResizeObserver")
  );
  return E.useEffect(() => {
    e && (o(e.itemHeight), n({ listHeight: 0, offsetTop: 0, visibleHeight: e.viewportHeight, visibleWidth: 100 }));
  }, [e, n, o]), /* @__PURE__ */ G("div", { "data-viewport-type": "window", ref: i, style: Je(!1, s), children: t });
}, Hr = /* @__PURE__ */ E.memo(function(e) {
  const n = M("useWindowScroll"), o = M("customScrollParent"), r = bt("fixedHeaderHeight"), s = bt("fixedFooterHeight"), i = M("fixedHeaderContent"), l = M("fixedFooterContent"), c = M("context"), d = Lt(
    E.useMemo(() => oe(r, (a) => Ht(a, "height")), [r]),
    !0,
    M("skipAnimationFrameInResizeObserver")
  ), m = Lt(
    E.useMemo(() => oe(s, (a) => Ht(a, "height")), [s]),
    !0,
    M("skipAnimationFrameInResizeObserver")
  ), S = o || n ? Br : Er, p = o || n ? Rr : br, T = M("TableComponent"), w = M("TableHeadComponent"), R = M("TableFooterComponent"), h = i ? /* @__PURE__ */ G(
    w,
    {
      ref: d,
      style: { position: "sticky", top: 0, zIndex: 2 },
      ...ot(w, c),
      children: i()
    },
    "TableHead"
  ) : null, f = l ? /* @__PURE__ */ G(
    R,
    {
      ref: m,
      style: { bottom: 0, position: "sticky", zIndex: 1 },
      ...ot(R, c),
      children: l()
    },
    "TableFoot"
  ) : null;
  return /* @__PURE__ */ G(S, { ...e, ...ot(S, c), children: /* @__PURE__ */ G(p, { children: /* @__PURE__ */ Dt(T, { style: { borderSpacing: 0, overflowAnchor: "none" }, ...ot(T, c), children: [
    h,
    /* @__PURE__ */ G(yr, {}, "TableBody"),
    f
  ] }) }) });
}), {
  Component: io,
  useEmitter: nn,
  useEmitterValue: M,
  usePublisher: bt
} = /* @__PURE__ */ Ye(
  Tr,
  {
    optional: {
      restoreStateFrom: "restoreStateFrom",
      context: "context",
      followOutput: "followOutput",
      firstItemIndex: "firstItemIndex",
      itemContent: "itemContent",
      groupContent: "groupContent",
      fixedHeaderContent: "fixedHeaderContent",
      fixedFooterContent: "fixedFooterContent",
      overscan: "overscan",
      increaseViewportBy: "increaseViewportBy",
      minOverscanItemCount: "minOverscanItemCount",
      totalCount: "totalCount",
      topItemCount: "topItemCount",
      initialTopMostItemIndex: "initialTopMostItemIndex",
      components: "components",
      groupCounts: "groupCounts",
      atBottomThreshold: "atBottomThreshold",
      atTopThreshold: "atTopThreshold",
      computeItemKey: "computeItemKey",
      defaultItemHeight: "defaultItemHeight",
      fixedGroupHeight: "fixedGroupHeight",
      // Must be set above 'fixedItemHeight'
      fixedItemHeight: "fixedItemHeight",
      itemSize: "itemSize",
      scrollSeekConfiguration: "scrollSeekConfiguration",
      data: "data",
      initialItemCount: "initialItemCount",
      initialScrollTop: "initialScrollTop",
      alignToBottom: "alignToBottom",
      useWindowScroll: "useWindowScroll",
      customScrollParent: "customScrollParent",
      scrollerRef: "scrollerRef",
      logLevel: "logLevel"
    },
    methods: {
      scrollToIndex: "scrollToIndex",
      scrollIntoView: "scrollIntoView",
      scrollTo: "scrollTo",
      scrollBy: "scrollBy",
      getState: "getState"
    },
    events: {
      isScrolling: "isScrolling",
      endReached: "endReached",
      startReached: "startReached",
      rangeChanged: "rangeChanged",
      atBottomStateChange: "atBottomStateChange",
      atTopStateChange: "atTopStateChange",
      totalListHeightChanged: "totalListHeightChanged",
      itemsRendered: "itemsRendered",
      groupIndices: "groupIndices"
    }
  },
  Hr
), Er = /* @__PURE__ */ Qe({ useEmitter: nn, useEmitterValue: M, usePublisher: bt }), Br = /* @__PURE__ */ tn({ useEmitter: nn, useEmitterValue: M, usePublisher: bt }), Xr = io, Jr = io, Rn = {
  bottom: 0,
  itemHeight: 0,
  items: [],
  itemWidth: 0,
  offsetBottom: 0,
  offsetTop: 0,
  top: 0
}, Or = {
  bottom: 0,
  itemHeight: 0,
  items: [{ index: 0 }],
  itemWidth: 0,
  offsetBottom: 0,
  offsetTop: 0,
  top: 0
}, { ceil: Hn, floor: ve, max: ne, min: ke, round: En } = Math;
function Bn(t, e, n) {
  return Array.from({ length: e - t + 1 }).map((o, r) => ({ data: n === null ? null : n[r + t], index: r + t }));
}
function Lr(t) {
  return {
    ...Or,
    items: t
  };
}
function Ie(t, e) {
  return t !== void 0 && t.width === e.width && t.height === e.height;
}
function kr(t, e) {
  return t !== void 0 && t.column === e.column && t.row === e.row;
}
const zr = /* @__PURE__ */ j(
  ([
    { increaseViewportBy: t, listBoundary: e, overscan: n, visibleRange: o },
    { footerHeight: r, headerHeight: s, scrollBy: i, scrollContainerState: l, scrollTo: c, scrollTop: d, smoothScrollTargetReached: m, viewportHeight: S },
    p,
    T,
    { didMount: w, propsReady: R },
    { customScrollParent: h, useWindowScroll: f, windowScrollContainerState: a, windowScrollTo: x, windowViewportRect: H },
    y
  ]) => {
    const L = v(0), u = v(0), g = v(Rn), C = v({ height: 0, width: 0 }), z = v({ height: 0, width: 0 }), O = U(), V = U(), N = v(0), Z = v(null), F = v({ column: 0, row: 0 }), mt = U(), q = U(), Q = v(!1), gt = v(0), ut = v(!0), St = v(!1), Ft = v(!1);
    Y(
      I(
        w,
        $(gt),
        P(([b, D]) => D !== 0)
      ),
      () => {
        _(ut, !1);
      }
    ), Y(
      I(
        at(w, ut, z, C, gt, St),
        P(([b, D, K, st, , tt]) => b && !D && K.height !== 0 && st.height !== 0 && !tt)
      ),
      ([, , , , b]) => {
        _(St, !0), Ue(1, () => {
          _(O, b);
        }), yt(I(d), () => {
          _(e, [0, 0]), _(ut, !0);
        });
      }
    ), k(
      I(
        q,
        P((b) => b != null && b.scrollTop > 0),
        Bt(0)
      ),
      u
    ), Y(
      I(
        w,
        $(q),
        P(([, b]) => b != null)
      ),
      ([, b]) => {
        b && (_(C, b.viewport), _(z, b.item), _(F, b.gap), b.scrollTop > 0 && (_(Q, !0), yt(I(d, $t(1)), (D) => {
          _(Q, !1);
        }), _(c, { top: b.scrollTop })));
      }
    ), k(
      I(
        C,
        B(({ height: b }) => b)
      ),
      S
    ), k(
      I(
        at(
          W(C, Ie),
          W(z, Ie),
          W(F, (b, D) => b !== void 0 && b.column === D.column && b.row === D.row),
          W(d)
        ),
        B(([b, D, K, st]) => ({
          gap: K,
          item: D,
          scrollTop: st,
          viewport: b
        }))
      ),
      mt
    ), k(
      I(
        at(
          W(L),
          o,
          W(F, kr),
          W(z, Ie),
          W(C, Ie),
          W(Z),
          W(u),
          W(Q),
          W(ut),
          W(gt)
        ),
        P(([, , , , , , , b]) => !b),
        B(
          ([
            b,
            [D, K],
            st,
            tt,
            X,
            lt,
            xt,
            ,
            ct,
            Vt
          ]) => {
            const { column: Wt, row: te } = st, { height: pe, width: He } = tt, { width: on } = X;
            if (xt === 0 && (b === 0 || on === 0))
              return Rn;
            if (He === 0) {
              const un = Ke(Vt, b), ao = un + Math.max(xt - 1, 0);
              return Lr(Bn(un, ao, lt));
            }
            const he = lo(on, He, Wt);
            let jt, Mt;
            ct ? D === 0 && K === 0 && xt > 0 ? (jt = 0, Mt = xt - 1) : (jt = he * ve((D + te) / (pe + te)), Mt = he * Hn((K + te) / (pe + te)) - 1, Mt = ke(b - 1, ne(Mt, he - 1)), jt = ke(Mt, ne(0, jt))) : (jt = 0, Mt = -1);
            const rn = Bn(jt, Mt, lt), { bottom: sn, top: ln } = On(X, st, tt, rn), cn = Hn(b / he), uo = cn * pe + (cn - 1) * te - sn;
            return { bottom: sn, itemHeight: pe, items: rn, itemWidth: He, offsetBottom: uo, offsetTop: ln, top: ln };
          }
        )
      ),
      g
    ), k(
      I(
        Z,
        P((b) => b !== null),
        B((b) => b.length)
      ),
      L
    ), k(
      I(
        at(C, z, g, F),
        P(([b, D, { items: K }]) => K.length > 0 && D.height !== 0 && b.height !== 0),
        B(([b, D, { items: K }, st]) => {
          const { bottom: tt, top: X } = On(b, st, D, K);
          return [X, tt];
        }),
        nt(ie)
      ),
      e
    );
    const pt = v(!1);
    k(
      I(
        d,
        $(pt),
        B(([b, D]) => D || b !== 0)
      ),
      pt
    );
    const Kt = vt(
      I(
        at(g, L),
        P(([{ items: b }]) => b.length > 0),
        $(pt),
        P(([[b, D], K]) => {
          const tt = b.items[b.items.length - 1].index === D - 1;
          return (K || b.bottom > 0 && b.itemHeight > 0 && b.offsetBottom === 0 && b.items.length === D) && tt;
        }),
        B(([[, b]]) => b - 1),
        nt()
      )
    ), Jt = vt(
      I(
        W(g),
        P(({ items: b }) => b.length > 0 && b[0].index === 0),
        Bt(0),
        nt()
      )
    ), Et = vt(
      I(
        W(g),
        $(Q),
        P(([{ items: b }, D]) => b.length > 0 && !D),
        B(([{ items: b }]) => ({
          endIndex: b[b.length - 1].index,
          startIndex: b[0].index
        })),
        nt(Mn),
        zt(0)
      )
    );
    k(Et, T.scrollSeekRangeChanged), k(
      I(
        O,
        $(C, z, L, F),
        B(([b, D, K, st, tt]) => {
          const X = Un(b), { align: lt, behavior: xt, offset: ct } = X;
          let Vt = X.index;
          Vt === "LAST" && (Vt = st - 1), Vt = ne(0, Vt, ke(st - 1, Vt));
          let Wt = Ge(D, tt, K, Vt);
          return lt === "end" ? Wt = En(Wt - D.height + K.height) : lt === "center" && (Wt = En(Wt - D.height / 2 + K.height / 2)), ct !== void 0 && ct !== 0 && (Wt += ct), { behavior: xt, top: Wt };
        })
      ),
      c
    );
    const Qt = ht(
      I(
        g,
        B((b) => b.offsetBottom + b.bottom)
      ),
      0
    );
    return k(
      I(
        H,
        B((b) => ({ height: b.visibleHeight, width: b.visibleWidth }))
      ),
      C
    ), {
      customScrollParent: h,
      // input
      data: Z,
      deviation: N,
      footerHeight: r,
      gap: F,
      headerHeight: s,
      increaseViewportBy: t,
      initialItemCount: u,
      itemDimensions: z,
      overscan: n,
      restoreStateFrom: q,
      scrollBy: i,
      scrollContainerState: l,
      scrollHeight: V,
      scrollTo: c,
      scrollToIndex: O,
      scrollTop: d,
      smoothScrollTargetReached: m,
      totalCount: L,
      useWindowScroll: f,
      viewportDimensions: C,
      windowScrollContainerState: a,
      windowScrollTo: x,
      windowViewportRect: H,
      ...T,
      // output
      gridState: g,
      horizontalDirection: Ft,
      initialTopMostItemIndex: gt,
      totalListHeight: Qt,
      ...p,
      endReached: Kt,
      propsReady: R,
      rangeChanged: Et,
      startReached: Jt,
      stateChanged: mt,
      stateRestoreInProgress: Q,
      ...y
    };
  },
  rt(je, It, me, Xn, At, qe, Gt)
);
function lo(t, e, n) {
  return ne(1, ve((t + n) / (ve(e) + n)));
}
function On(t, e, n, o) {
  const { height: r } = n;
  if (r === void 0 || o.length === 0)
    return { bottom: 0, top: 0 };
  const s = Ge(t, e, n, o[0].index);
  return { bottom: Ge(t, e, n, o[o.length - 1].index) + r, top: s };
}
function Ge(t, e, n, o) {
  const r = lo(t.width, n.width, e.column), s = ve(o / r), i = s * n.height + ne(0, s - 1) * e.row;
  return i > 0 ? i + e.row : i;
}
const Fr = /* @__PURE__ */ j(() => {
  const t = v((S) => `Item ${S}`), e = v({}), n = v(null), o = v("virtuoso-grid-item"), r = v("virtuoso-grid-list"), s = v(Xe), i = v("div"), l = v(Xt), c = (S, p = null) => ht(
    I(
      e,
      B((T) => T[S]),
      nt()
    ),
    p
  ), d = v(!1), m = v(!1);
  return k(W(m), d), {
    components: e,
    computeItemKey: s,
    context: n,
    FooterComponent: c("Footer"),
    HeaderComponent: c("Header"),
    headerFooterTag: i,
    itemClassName: o,
    ItemComponent: c("Item", "div"),
    itemContent: t,
    listClassName: r,
    ListComponent: c("List", "div"),
    readyStateChanged: d,
    reportReadyState: m,
    ScrollerComponent: c("Scroller", "div"),
    scrollerRef: l,
    ScrollSeekPlaceholder: c("ScrollSeekPlaceholder", "div")
  };
}), Vr = /* @__PURE__ */ j(
  ([t, e]) => ({ ...t, ...e }),
  rt(zr, Fr)
), Wr = /* @__PURE__ */ E.memo(function() {
  const e = et("gridState"), n = et("listClassName"), o = et("itemClassName"), r = et("itemContent"), s = et("computeItemKey"), i = et("isSeeking"), l = wt("scrollHeight"), c = et("ItemComponent"), d = et("ListComponent"), m = et("ScrollSeekPlaceholder"), S = et("context"), p = wt("itemDimensions"), T = wt("gap"), w = et("log"), R = et("stateRestoreInProgress"), h = wt("reportReadyState"), f = Lt(
    E.useMemo(
      () => (a) => {
        const x = a.parentElement.parentElement.scrollHeight;
        l(x);
        const H = a.firstChild;
        if (H !== null) {
          const { height: y, width: L } = H.getBoundingClientRect();
          p({ height: y, width: L });
        }
        T({
          column: Ln("column-gap", getComputedStyle(a).columnGap, w),
          row: Ln("row-gap", getComputedStyle(a).rowGap, w)
        });
      },
      [l, p, T, w]
    ),
    !0,
    !1
  );
  return eo(() => {
    e.itemHeight > 0 && e.itemWidth > 0 && h(!0);
  }, [e]), R ? null : /* @__PURE__ */ G(
    d,
    {
      className: n,
      ref: f,
      ...ot(d, S),
      "data-testid": "virtuoso-item-list",
      style: { paddingBottom: e.offsetBottom, paddingTop: e.offsetTop },
      children: e.items.map((a) => {
        const x = s(a.index, a.data, S);
        return i ? /* @__PURE__ */ G(
          m,
          {
            ...ot(m, S),
            height: e.itemHeight,
            index: a.index,
            width: e.itemWidth
          },
          x
        ) : /* @__PURE__ */ Nt(
          c,
          {
            ...ot(c, S),
            className: o,
            "data-index": a.index,
            key: x
          },
          r(a.index, a.data, S)
        );
      })
    }
  );
}), Pr = E.memo(function() {
  const e = et("HeaderComponent"), n = wt("headerHeight"), o = et("headerFooterTag"), r = Lt(
    E.useMemo(
      () => (i) => {
        n(Ht(i, "height"));
      },
      [n]
    ),
    !0,
    !1
  ), s = et("context");
  return e != null ? /* @__PURE__ */ G(o, { ref: r, children: /* @__PURE__ */ G(e, { ...ot(e, s) }) }) : null;
}), Gr = E.memo(function() {
  const e = et("FooterComponent"), n = wt("footerHeight"), o = et("headerFooterTag"), r = Lt(
    E.useMemo(
      () => (i) => {
        n(Ht(i, "height"));
      },
      [n]
    ),
    !0,
    !1
  ), s = et("context");
  return e != null ? /* @__PURE__ */ G(o, { ref: r, children: /* @__PURE__ */ G(e, { ...ot(e, s) }) }) : null;
}), Ar = ({ children: t }) => {
  const e = E.useContext(to), n = wt("itemDimensions"), o = wt("viewportDimensions"), r = Lt(
    E.useMemo(
      () => (s) => {
        o(s.getBoundingClientRect());
      },
      [o]
    ),
    !0,
    !1
  );
  return E.useEffect(() => {
    e && (o({ height: e.viewportHeight, width: e.viewportWidth }), n({ height: e.itemHeight, width: e.itemWidth }));
  }, [e, o, n]), /* @__PURE__ */ G("div", { ref: r, style: Re(!1), children: t });
}, Mr = ({ children: t }) => {
  const e = E.useContext(to), n = wt("windowViewportRect"), o = wt("itemDimensions"), r = et("customScrollParent"), s = et("useWindowScroll"), i = Ne(n, r, !1);
  return E.useEffect(() => {
    e && (o({ height: e.itemHeight, width: e.itemWidth }), n({ listHeight: 0, offsetTop: 0, visibleHeight: e.viewportHeight, visibleWidth: e.viewportWidth }));
  }, [e, n, o]), /* @__PURE__ */ G("div", { ref: i, style: Je(!1, s), children: t });
}, _r = /* @__PURE__ */ E.memo(function({ ...e }) {
  const n = et("useWindowScroll"), o = et("customScrollParent"), r = o || n ? $r : Dr, s = o || n ? Mr : Ar, i = et("context");
  return /* @__PURE__ */ G(r, { ...e, ...ot(r, i), children: /* @__PURE__ */ Dt(s, { children: [
    /* @__PURE__ */ G(Pr, {}),
    /* @__PURE__ */ G(Wr, {}),
    /* @__PURE__ */ G(Gr, {})
  ] }) });
}), {
  Component: Nr,
  useEmitter: co,
  useEmitterValue: et,
  usePublisher: wt
} = /* @__PURE__ */ Ye(
  Vr,
  {
    optional: {
      context: "context",
      totalCount: "totalCount",
      overscan: "overscan",
      itemContent: "itemContent",
      components: "components",
      computeItemKey: "computeItemKey",
      data: "data",
      initialItemCount: "initialItemCount",
      scrollSeekConfiguration: "scrollSeekConfiguration",
      headerFooterTag: "headerFooterTag",
      listClassName: "listClassName",
      itemClassName: "itemClassName",
      useWindowScroll: "useWindowScroll",
      customScrollParent: "customScrollParent",
      scrollerRef: "scrollerRef",
      logLevel: "logLevel",
      restoreStateFrom: "restoreStateFrom",
      initialTopMostItemIndex: "initialTopMostItemIndex",
      increaseViewportBy: "increaseViewportBy"
    },
    methods: {
      scrollTo: "scrollTo",
      scrollBy: "scrollBy",
      scrollToIndex: "scrollToIndex"
    },
    events: {
      isScrolling: "isScrolling",
      endReached: "endReached",
      startReached: "startReached",
      rangeChanged: "rangeChanged",
      atBottomStateChange: "atBottomStateChange",
      atTopStateChange: "atTopStateChange",
      stateChanged: "stateChanged",
      readyStateChanged: "readyStateChanged"
    }
  },
  _r
), Dr = /* @__PURE__ */ Qe({ useEmitter: co, useEmitterValue: et, usePublisher: wt }), $r = /* @__PURE__ */ tn({ useEmitter: co, useEmitterValue: et, usePublisher: wt });
function Ln(t, e, n) {
  return e !== "normal" && e?.endsWith("px") !== !0 && n(`${t} was not resolved to pixel value correctly`, e, ft.WARN), e === "normal" ? 0 : parseInt(e ?? "0", 10);
}
const Qr = Nr;
export {
  Jr as GroupedTableVirtuoso,
  Zr as GroupedVirtuoso,
  ft as LogLevel,
  Xr as TableVirtuoso,
  Yr as Virtuoso,
  Qr as VirtuosoGrid,
  to as VirtuosoGridMockContext,
  be as VirtuosoMockContext
};
